// openclaw/dist/hooks-cli-CRFCvPRq.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import { r as theme } from "./theme-CdOoMzRk.js";
import "./globals-DBUMOBZ8.js";
import { m as defaultRuntime } from "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import { S as shortenHomePath } from "./utils-CS0Ikux6.js";
import { t as formatDocsLink } from "./links-8xRhWBQL.js";
import "./paths-BFl2-hCf.js";
import "./auth-profiles-CVFpgeL-.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import { m as resolveDefaultAgentId, p as resolveAgentWorkspaceDir } from "./agent-scope-DoT9OqaV.js";
import "./model-selection-B6ao45a4.js";
import { g as writeConfigFile, s as loadConfig } from "./io-cPs4dU7X.js";
import "./host-env-security-DEKL50zA.js";
import "./shell-env-BKuVS72k.js";
import "./safe-text-DIDDfQyI.js";
import "./version-BpHNkJed.js";
import "./env-substitution-B8NFl2Jd.js";
import "./includes-CCK6fRRs.js";
import "./zod-schema.providers-core-DLPfih2y.js";
import "./legacy-web-search-CTvir-Bl.js";
import "./registry-xyHjVLxh.js";
import "./config-state-Dx9-tLPS.js";
import "./min-host-version-xbc6BJ_K.js";
import "./manifest-registry-D5E7Gxgl.js";
import "./runtime-guard-QGt7fm0l.js";
import "./avatar-policy-DVf9B9eu.js";
import "./ip-DIgtRRTW.js";
import "./zod-schema.agent-runtime-C659GPl8.js";
import "./zod-schema.core-yLTNC4-K.js";
import "./config-CB4aYWqd.js";
import "./file-lock-DoKDW8jx.js";
import "./audit-fs-CEN00XrG.js";
import "./resolve-uXRpJb-M.js";
import "./profiles-Cyg58FCO.js";
import { t as formatCliCommand } from "./command-format-DRptdHvm.js";
import "./tailscale-CSUwCuE9.js";
import "./tailnet-BVzEE6AW.js";
import "./net-Dk658jWW.js";
import "./auth-C_pTuZtn.js";
import "./credentials-CO55Yx_u.js";
import "./message-channel-DUrzQUcI.js";
import "./store-D_eNuCKK.js";
import "./runtime-C8dQugND.js";
import "./plugins-DGdgUNcN.js";
import "./sessions-C8WesgCl.js";
import "./paths-BPfguEtH.js";
import "./session-write-lock-bE2vQvvO.js";
import "./method-scopes-BWG4Q18M.js";
import "./call-VZCb020X.js";
import "./prompt-style-MhFLSlua.js";
import "./ports-lsof-C3NOmuKA.js";
import "./restart-stale-pids-CQbYk99d.js";
import "./ports-CQIuVpXl.js";
import "./logging-C2L37N2X.js";
import "./commands-CrPoYX9r.js";
import "./issue-format-CP-gqjZB.js";
import "./identity-D8H54Ni5.js";
import "./heartbeat-FZaB85hC.js";
import "./pi-embedded-CzQCqSlH.js";
import "./internal-hooks-D4lZfNM5.js";
import "./multimodal-De3qv_72.js";
import "./memory-search-DRm4OVlH.js";
import "./provider-catalog-7bUKai5_.js";
import "./secret-input-BNBMqe2K.js";
import "./bindings-RcwGGOd3.js";
import "./resolve-route-DbP_XX4E.js";
import "./routing-B37UwRwq.js";
import "./identity-file-CCks_qJo.js";
import "./outbound-runtime-oDjCAP7B.js";
import "./provider-env-vars-Bhzl_gIs.js";
import "./provider-auth-input-DxjPWsSV.js";
import "./provider-model-minimax-XdwKeUOE.js";
import "./provider-models-DfgBoVpf.js";
import "./anthropic-vertex-provider-e_sRR8fp.js";
import "./models-config.providers.discovery-Ch23kOpG.js";
import "./text-runtime-B9_24WY5.js";
import "./tool-catalog-6VPSKPp9.js";
import "./docker-YL21-mX4.js";
import "./sandbox-C3lR2qEl.js";
import "./common-De-BVwIv.js";
import "./image-ops-BLXW6hKm.js";
import "./thinking-CSfcimRZ.js";
import "./path-alias-guards-CgP-6a1d.js";
import "./sandbox-paths-Bk2ZwMsR.js";
import "./channel-actions-BzT1SAu9.js";
import "./mime-D7Q9o2pi.js";
import "./ssrf-rHmQXFZ4.js";
import "./fetch-guard-BH3MjZoA.js";
import "./provider-web-search-DYdDI2c9.js";
import "./manager-CTkU8M9E.js";
import "./config-schema-Bpwy_blm.js";
import "./channel-plugin-common-COvcxNrG.js";
import "./status-helpers-Dd-hr4j4.js";
import "./conversation-runtime-CqOm9f5F.js";
import "./runtime-whatsapp-boundary-CWceJs-_.js";
import "./pairing-store-ClZkWB4v.js";
import "./json-store-Ju23bywi.js";
import "./secret-file-Ctk8Gt6Y.js";
import "./channel-config-schema-DGuU5OU5.js";
import "./setup-binary-DXdzcWG_.js";
import "./archive-Bk8HKpaY.js";
import "./fs-safe-Dvzkbqib.js";
import "./signal-cli-install-BJB9N18P.js";
import "./setup-wizard-proxy-D0W3ulU0.js";
import "./setup-BYrV6xB7.js";
import "./dm-policy-shared-DSMJu_ZS.js";
import "./hook-runtime-DEbp5SYL.js";
import "./diagnostic-PF4suwLJ.js";
import "./templating-BueU7auS.js";
import "./channel-reply-pipeline-C0iCnI2R.js";
import "./reply-history-BVE7m0XA.js";
import "./commands-registry.data-Dr5wKe-o.js";
import "./commands-registry-C-vLUx3f.js";
import "./frontmatter-DUZeCF_V.js";
import "./env-overrides-DbK0fm-W.js";
import "./skills-CreXENx9.js";
import "./skills-remote-Ctc3qkkj.js";
import "./workspace-dirs-CKrvld61.js";
import "./pairing-token-CT1j20jA.js";
import "./skill-commands-DDonxE3L.js";
import "./level-overrides-DoL5t4pG.js";
import "./config-BkPtspys.js";
import "./routes-CdaHRCET.js";
import "./ssh-tunnel-CnI7n7iv.js";
import "./server-middleware-DcKqo5DB.js";
import "./logging-CSCaAwMS.js";
import "./config-runtime-ST43sCMq.js";
import "./exec-approvals-CVzy3Vfp.js";
import "./webhook-ingress-Chu9Ldhy.js";
import "./system-events-D9mayVHC.js";
import "./ssrf-policy-C4n9geuA.js";
import "./provider-auth-ref-D7iRHU8I.js";
import "./provider-auth-helpers-BeqTsaJO.js";
import "./provider-api-key-auth-C6RcLhw0.js";
import "./runtime-env-CjqWVeEh.js";
import "./pairing-labels-BOG3E444.js";
import "./directory-runtime-Dv9ukfmB.js";
import "./read-only-account-inspect-gnHd6Zsr.js";
import "./web-media-C4M86664.js";
import "./temp-path-BPp7EkAA.js";
import "./media-understanding-BGz4x5yF.js";
import "./web-media-ChNoP5pI.js";
import "./state-paths-kGWIOBBS.js";
import "./llm-task-CVbsbiq5.js";
import "./pi-model-discovery-BpWH93Dy.js";
import "./exec-inline-eval-BXBKoGM8.js";
import "./target-registry-Che-i6sL.js";
import "./external-content-DD7F5dje.js";
import "./brave-DlJT7GTf.js";
import "./duckduckgo-Bkn2wdAA.js";
import "./exa-DpNIw7fA.js";
import "./security-runtime-kOJWyLps.js";
import "./provider-usage-BWl6XLRe.js";
import "./provider-onboard-DSTfYPeq.js";
import "./perplexity-Diz33fcS.js";
import "./stagger-BkeVxPT8.js";
import "./command-secret-targets-BYcKChcy.js";
import "./delivery-queue-gz62qXSE.js";
import "./channel-summary-wFTF6Nzp.js";
import "./session-system-events-CXg2CZLB.js";
import "./tool-policy-match-BNFjIlBI.js";
import "./runtime-DB0HL3pI.js";
import "./channel-status-BsYjzyR-.js";
import "./discord-core-bSS6QgzK.js";
import "./cli-runtime-hF9d0wpB.js";
import "./config-presence-CovoxGn8.js";
import "./query-expansion-CzcRNQqC.js";
import "./search-manager-PSAQEKfx.js";
import "./acp-runtime-BUMdU4AF.js";
import "./telegram-core-BaYrcVUp.js";
import "./audit-CO95QNfX.js";
import "./gateway-runtime-DhcHrysu.js";
import "./connection-auth-COroqkf4.js";
import "./mcp-config-Dp646rw6.js";
import "./frontmatter-CJE7MBD1.js";
import { a as resolveHookEntries } from "./config-CbdGijQT.js";
import "./install-source-utils-L4jI7Syn.js";
import "./plugins-install-persist-daoXRUwr.js";
import "./clawhub-xVLMmaOX.js";
import "./installs-XLSjUYkq.js";
import "./clawhub-BprlOTz-.js";
import { a as buildPluginStatusReport } from "./status-DE_JvlNY.js";
import "./plugin-install-plan-BSvG_fXq.js";
import "./bundled-sources-nznCMKgA.js";
import { n as renderTable, t as getTerminalTableWidth } from "./table-D77MmWwB.js";
import { t as loadWorkspaceHookEntries } from "./workspace-Mvv67V6q.js";
import { t as buildWorkspaceHookStatus } from "./hooks-status-C03feaA_.js";
import { n as runPluginInstallCommand, t as runPluginUpdateCommand } from "./plugins-update-command-D6mIghfk.js";
import "./update-WJLFhMS8.js";
//#region src/cli/hooks-cli.ts
function mergeHookEntries(pluginEntries, workspaceEntries) {
	return resolveHookEntries([...pluginEntries, ...workspaceEntries]);
}
function buildHooksReport(config) {
	const workspaceDir = resolveAgentWorkspaceDir(config, resolveDefaultAgentId(config));
	const workspaceEntries = loadWorkspaceHookEntries(workspaceDir, { config });
	return buildWorkspaceHookStatus(workspaceDir, {
		config,
		entries: mergeHookEntries(buildPluginStatusReport({
			config,
			workspaceDir
		}).hooks.map((hook) => hook.entry), workspaceEntries)
	});
}
function resolveHookForToggle(report, hookName, opts) {
	const hook = report.hooks.find((h) => h.name === hookName);
	if (!hook) throw new Error(`Hook "${hookName}" not found`);
	if (hook.managedByPlugin) throw new Error(`Hook "${hookName}" is managed by plugin "${hook.pluginId ?? "unknown"}" and cannot be enabled/disabled.`);
	if (opts?.requireEligible && !hook.requirementsSatisfied) throw new Error(`Hook "${hookName}" is not eligible (missing requirements)`);
	return hook;
}
function buildConfigWithHookEnabled(params) {
	const entries = { ...params.config.hooks?.internal?.entries };
	entries[params.hookName] = {
		...entries[params.hookName],
		enabled: params.enabled
	};
	const internal = {
		...params.config.hooks?.internal,
		...params.ensureHooksEnabled ? { enabled: true } : {},
		entries
	};
	return {
		...params.config,
		hooks: {
			...params.config.hooks,
			internal
		}
	};
}
function formatHookStatus(hook) {
	if (hook.loadable) return theme.success("✓ ready");
	if (!hook.enabledByConfig) return theme.warn("⏸ disabled");
	return theme.error("✗ missing");
}
function formatHookName(hook) {
	return `${hook.emoji ?? "🔗"} ${theme.command(hook.name)}`;
}
function formatHookSource(hook) {
	if (!hook.managedByPlugin) return hook.source;
	return `plugin:${hook.pluginId ?? "unknown"}`;
}
function formatHookMissingSummary(hook) {
	const missing = [];
	if (hook.missing.bins.length > 0) missing.push(`bins: ${hook.missing.bins.join(", ")}`);
	if (hook.missing.anyBins.length > 0) missing.push(`anyBins: ${hook.missing.anyBins.join(", ")}`);
	if (hook.missing.env.length > 0) missing.push(`env: ${hook.missing.env.join(", ")}`);
	if (hook.missing.config.length > 0) missing.push(`config: ${hook.missing.config.join(", ")}`);
	if (hook.missing.os.length > 0) missing.push(`os: ${hook.missing.os.join(", ")}`);
	return missing.join("; ");
}
function exitHooksCliWithError(err) {
	defaultRuntime.error(`${theme.error("Error:")} ${err instanceof Error ? err.message : String(err)}`);
	process.exit(1);
}
function writeHooksOutput(value, json) {
	if (json) {
		defaultRuntime.writeStdout(value);
		return;
	}
	defaultRuntime.log(value);
}
async function runHooksCliAction(action) {
	try {
		await action();
	} catch (err) {
		exitHooksCliWithError(err);
	}
}
/**
* Format the hooks list output
*/
function formatHooksList(report, opts) {
	const hooks = opts.eligible ? report.hooks.filter((h) => h.loadable) : report.hooks;
	if (opts.json) {
		const jsonReport = {
			workspaceDir: report.workspaceDir,
			managedHooksDir: report.managedHooksDir,
			hooks: hooks.map((h) => ({
				name: h.name,
				description: h.description,
				emoji: h.emoji,
				eligible: h.loadable,
				disabled: !h.enabledByConfig,
				enabledByConfig: h.enabledByConfig,
				requirementsSatisfied: h.requirementsSatisfied,
				loadable: h.loadable,
				blockedReason: h.blockedReason,
				source: h.source,
				pluginId: h.pluginId,
				events: h.events,
				homepage: h.homepage,
				missing: h.missing,
				managedByPlugin: h.managedByPlugin
			}))
		};
		return JSON.stringify(jsonReport, null, 2);
	}
	if (hooks.length === 0) return opts.eligible ? `No eligible hooks found. Run \`${formatCliCommand("openclaw hooks list")}\` to see all hooks.` : "No hooks found.";
	const eligible = hooks.filter((h) => h.loadable);
	const tableWidth = getTerminalTableWidth();
	const rows = hooks.map((hook) => {
		const missing = formatHookMissingSummary(hook);
		return {
			Status: formatHookStatus(hook),
			Hook: formatHookName(hook),
			Description: theme.muted(hook.description),
			Source: formatHookSource(hook),
			Missing: missing ? theme.warn(missing) : ""
		};
	});
	const columns = [
		{
			key: "Status",
			header: "Status",
			minWidth: 10
		},
		{
			key: "Hook",
			header: "Hook",
			minWidth: 18,
			flex: true
		},
		{
			key: "Description",
			header: "Description",
			minWidth: 24,
			flex: true
		},
		{
			key: "Source",
			header: "Source",
			minWidth: 12,
			flex: true
		}
	];
	if (opts.verbose) columns.push({
		key: "Missing",
		header: "Missing",
		minWidth: 18,
		flex: true
	});
	const lines = [];
	lines.push(`${theme.heading("Hooks")} ${theme.muted(`(${eligible.length}/${hooks.length} ready)`)}`);
	lines.push(renderTable({
		width: tableWidth,
		columns,
		rows
	}).trimEnd());
	return lines.join("\n");
}
/**
* Format detailed info for a single hook
*/
function formatHookInfo(report, hookName, opts) {
	const hook = report.hooks.find((h) => h.name === hookName || h.hookKey === hookName);
	if (!hook) {
		if (opts.json) return JSON.stringify({
			error: "not found",
			hook: hookName
		}, null, 2);
		return `Hook "${hookName}" not found. Run \`${formatCliCommand("openclaw hooks list")}\` to see available hooks.`;
	}
	if (opts.json) return JSON.stringify({
		...hook,
		eligible: hook.loadable,
		disabled: !hook.enabledByConfig
	}, null, 2);
	const lines = [];
	const emoji = hook.emoji ?? "🔗";
	const status = hook.loadable ? theme.success("✓ Ready") : !hook.enabledByConfig ? theme.warn("⏸ Disabled") : theme.error("✗ Missing requirements");
	lines.push(`${emoji} ${theme.heading(hook.name)} ${status}`);
	lines.push("");
	lines.push(hook.description);
	lines.push("");
	lines.push(theme.heading("Details:"));
	if (hook.managedByPlugin) lines.push(`${theme.muted("  Source:")} ${hook.source} (${hook.pluginId ?? "unknown"})`);
	else lines.push(`${theme.muted("  Source:")} ${hook.source}`);
	lines.push(`${theme.muted("  Path:")} ${shortenHomePath(hook.filePath)}`);
	lines.push(`${theme.muted("  Handler:")} ${shortenHomePath(hook.handlerPath)}`);
	if (hook.homepage) lines.push(`${theme.muted("  Homepage:")} ${hook.homepage}`);
	if (hook.events.length > 0) lines.push(`${theme.muted("  Events:")} ${hook.events.join(", ")}`);
	if (hook.managedByPlugin) lines.push(theme.muted("  Managed by plugin; enable/disable via hooks CLI not available."));
	if (hook.blockedReason) lines.push(`${theme.muted("  Blocked reason:")} ${hook.blockedReason}`);
	if (hook.requirements.bins.length > 0 || hook.requirements.anyBins.length > 0 || hook.requirements.env.length > 0 || hook.requirements.config.length > 0 || hook.requirements.os.length > 0) {
		lines.push("");
		lines.push(theme.heading("Requirements:"));
		if (hook.requirements.bins.length > 0) {
			const binsStatus = hook.requirements.bins.map((bin) => {
				return hook.missing.bins.includes(bin) ? theme.error(`✗ ${bin}`) : theme.success(`✓ ${bin}`);
			});
			lines.push(`${theme.muted("  Binaries:")} ${binsStatus.join(", ")}`);
		}
		if (hook.requirements.anyBins.length > 0) {
			const anyBinsStatus = hook.missing.anyBins.length > 0 ? theme.error(`✗ (any of: ${hook.requirements.anyBins.join(", ")})`) : theme.success(`✓ (any of: ${hook.requirements.anyBins.join(", ")})`);
			lines.push(`${theme.muted("  Any binary:")} ${anyBinsStatus}`);
		}
		if (hook.requirements.env.length > 0) {
			const envStatus = hook.requirements.env.map((env) => {
				return hook.missing.env.includes(env) ? theme.error(`✗ ${env}`) : theme.success(`✓ ${env}`);
			});
			lines.push(`${theme.muted("  Environment:")} ${envStatus.join(", ")}`);
		}
		if (hook.requirements.config.length > 0) {
			const configStatus = hook.configChecks.map((check) => {
				return check.satisfied ? theme.success(`✓ ${check.path}`) : theme.error(`✗ ${check.path}`);
			});
			lines.push(`${theme.muted("  Config:")} ${configStatus.join(", ")}`);
		}
		if (hook.requirements.os.length > 0) {
			const osStatus = hook.missing.os.length > 0 ? theme.error(`✗ (${hook.requirements.os.join(", ")})`) : theme.success(`✓ (${hook.requirements.os.join(", ")})`);
			lines.push(`${theme.muted("  OS:")} ${osStatus}`);
		}
	}
	return lines.join("\n");
}
/**
* Format check output
*/
function formatHooksCheck(report, opts) {
	if (opts.json) {
		const eligible = report.hooks.filter((h) => h.loadable);
		const notEligible = report.hooks.filter((h) => !h.loadable);
		return JSON.stringify({
			total: report.hooks.length,
			eligible: eligible.length,
			notEligible: notEligible.length,
			hooks: {
				eligible: eligible.map((h) => h.name),
				notEligible: notEligible.map((h) => ({
					name: h.name,
					blockedReason: h.blockedReason,
					missing: h.missing
				}))
			}
		}, null, 2);
	}
	const eligible = report.hooks.filter((h) => h.loadable);
	const notEligible = report.hooks.filter((h) => !h.loadable);
	const lines = [];
	lines.push(theme.heading("Hooks Status"));
	lines.push("");
	lines.push(`${theme.muted("Total hooks:")} ${report.hooks.length}`);
	lines.push(`${theme.success("Ready:")} ${eligible.length}`);
	lines.push(`${theme.warn("Not ready:")} ${notEligible.length}`);
	if (notEligible.length > 0) {
		lines.push("");
		lines.push(theme.heading("Hooks not ready:"));
		for (const hook of notEligible) {
			const reasons = [];
			if (hook.blockedReason && hook.blockedReason !== "missing requirements") reasons.push(hook.blockedReason);
			if (hook.missing.bins.length > 0) reasons.push(`bins: ${hook.missing.bins.join(", ")}`);
			if (hook.missing.anyBins.length > 0) reasons.push(`anyBins: ${hook.missing.anyBins.join(", ")}`);
			if (hook.missing.env.length > 0) reasons.push(`env: ${hook.missing.env.join(", ")}`);
			if (hook.missing.config.length > 0) reasons.push(`config: ${hook.missing.config.join(", ")}`);
			if (hook.missing.os.length > 0) reasons.push(`os: ${hook.missing.os.join(", ")}`);
			lines.push(`  ${hook.emoji ?? "🔗"} ${hook.name} - ${reasons.join("; ")}`);
		}
	}
	return lines.join("\n");
}
async function enableHook(hookName) {
	const config = loadConfig();
	const hook = resolveHookForToggle(buildHooksReport(config), hookName, { requireEligible: true });
	await writeConfigFile(buildConfigWithHookEnabled({
		config,
		hookName,
		enabled: true,
		ensureHooksEnabled: true
	}));
	defaultRuntime.log(`${theme.success("✓")} Enabled hook: ${hook.emoji ?? "🔗"} ${theme.command(hookName)}`);
}
async function disableHook(hookName) {
	const config = loadConfig();
	const hook = resolveHookForToggle(buildHooksReport(config), hookName);
	await writeConfigFile(buildConfigWithHookEnabled({
		config,
		hookName,
		enabled: false
	}));
	defaultRuntime.log(`${theme.warn("⏸")} Disabled hook: ${hook.emoji ?? "🔗"} ${theme.command(hookName)}`);
}
function registerHooksCli(program) {
	const hooks = program.command("hooks").description("Manage internal agent hooks").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/hooks", "docs.openclaw.ai/cli/hooks")}\n`);
	hooks.command("list").description("List all hooks").option("--eligible", "Show only eligible hooks", false).option("--json", "Output as JSON", false).option("-v, --verbose", "Show more details including missing requirements", false).action(async (opts) => runHooksCliAction(async () => {
		writeHooksOutput(formatHooksList(buildHooksReport(loadConfig()), opts), opts.json);
	}));
	hooks.command("info <name>").description("Show detailed information about a hook").option("--json", "Output as JSON", false).action(async (name, opts) => runHooksCliAction(async () => {
		writeHooksOutput(formatHookInfo(buildHooksReport(loadConfig()), name, opts), opts.json);
	}));
	hooks.command("check").description("Check hooks eligibility status").option("--json", "Output as JSON", false).action(async (opts) => runHooksCliAction(async () => {
		writeHooksOutput(formatHooksCheck(buildHooksReport(loadConfig()), opts), opts.json);
	}));
	hooks.command("enable <name>").description("Enable a hook").action(async (name) => runHooksCliAction(async () => {
		await enableHook(name);
	}));
	hooks.command("disable <name>").description("Disable a hook").action(async (name) => runHooksCliAction(async () => {
		await disableHook(name);
	}));
	hooks.command("install").description("Deprecated: install a hook pack via `openclaw plugins install`").argument("<path-or-spec>", "Path to a hook pack or npm package spec").option("-l, --link", "Link a local path instead of copying", false).option("--pin", "Record npm installs as exact resolved <name>@<version>", false).action(async (raw, opts) => {
		defaultRuntime.log(theme.warn("`openclaw hooks install` is deprecated; use `openclaw plugins install`."));
		await runPluginInstallCommand({
			raw,
			opts
		});
	});
	hooks.command("update").description("Deprecated: update hook packs via `openclaw plugins update`").argument("[id]", "Hook pack id (omit with --all)").option("--all", "Update all tracked hooks", false).option("--dry-run", "Show what would change without writing", false).action(async (id, opts) => {
		defaultRuntime.log(theme.warn("`openclaw hooks update` is deprecated; use `openclaw plugins update`."));
		await runPluginUpdateCommand({
			id,
			opts
		});
	});
	hooks.action(async () => runHooksCliAction(async () => {
		const report = buildHooksReport(loadConfig());
		defaultRuntime.log(formatHooksList(report, {}));
	}));
}
//#endregion
export { registerHooksCli };
