// openclaw/dist/models-cli-b943eUPk.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import { r as theme } from "./theme-CdOoMzRk.js";
import "./globals-DBUMOBZ8.js";
import { m as defaultRuntime } from "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import "./utils-CS0Ikux6.js";
import { t as formatDocsLink } from "./links-8xRhWBQL.js";
import "./paths-BFl2-hCf.js";
import "./auth-profiles-CVFpgeL-.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import "./agent-scope-DoT9OqaV.js";
import "./model-selection-B6ao45a4.js";
import "./io-cPs4dU7X.js";
import "./host-env-security-DEKL50zA.js";
import "./shell-env-BKuVS72k.js";
import "./safe-text-DIDDfQyI.js";
import "./version-BpHNkJed.js";
import "./env-substitution-B8NFl2Jd.js";
import "./includes-CCK6fRRs.js";
import "./zod-schema.providers-core-DLPfih2y.js";
import "./legacy-web-search-CTvir-Bl.js";
import "./registry-xyHjVLxh.js";
import "./config-state-Dx9-tLPS.js";
import "./min-host-version-xbc6BJ_K.js";
import "./manifest-registry-D5E7Gxgl.js";
import "./runtime-guard-QGt7fm0l.js";
import "./avatar-policy-DVf9B9eu.js";
import "./ip-DIgtRRTW.js";
import "./zod-schema.agent-runtime-C659GPl8.js";
import "./zod-schema.core-yLTNC4-K.js";
import "./config-CB4aYWqd.js";
import "./file-lock-DoKDW8jx.js";
import "./audit-fs-CEN00XrG.js";
import "./resolve-uXRpJb-M.js";
import "./profiles-Cyg58FCO.js";
import "./tailscale-CSUwCuE9.js";
import "./tailnet-BVzEE6AW.js";
import "./net-Dk658jWW.js";
import "./auth-C_pTuZtn.js";
import "./credentials-CO55Yx_u.js";
import "./message-channel-DUrzQUcI.js";
import "./store-D_eNuCKK.js";
import "./runtime-C8dQugND.js";
import "./plugins-DGdgUNcN.js";
import "./sessions-C8WesgCl.js";
import "./paths-BPfguEtH.js";
import "./session-write-lock-bE2vQvvO.js";
import "./method-scopes-BWG4Q18M.js";
import "./call-VZCb020X.js";
import "./control-ui-shared-SxKiMaO4.js";
import "./onboard-helpers-Cks5EPpw.js";
import "./prompt-style-MhFLSlua.js";
import "./ports-lsof-C3NOmuKA.js";
import "./restart-stale-pids-CQbYk99d.js";
import "./ports-CQIuVpXl.js";
import "./logging-C2L37N2X.js";
import "./commands-CrPoYX9r.js";
import "./issue-format-CP-gqjZB.js";
import "./progress-CWxCPcAp.js";
import "./identity-D8H54Ni5.js";
import "./heartbeat-FZaB85hC.js";
import "./pi-embedded-CzQCqSlH.js";
import "./internal-hooks-D4lZfNM5.js";
import "./multimodal-De3qv_72.js";
import "./memory-search-DRm4OVlH.js";
import "./provider-catalog-7bUKai5_.js";
import "./secret-input-BNBMqe2K.js";
import "./bindings-RcwGGOd3.js";
import "./resolve-route-DbP_XX4E.js";
import "./routing-B37UwRwq.js";
import "./identity-file-CCks_qJo.js";
import "./outbound-runtime-oDjCAP7B.js";
import "./provider-env-vars-Bhzl_gIs.js";
import "./provider-auth-input-DxjPWsSV.js";
import "./provider-model-minimax-XdwKeUOE.js";
import "./provider-models-DfgBoVpf.js";
import "./anthropic-vertex-provider-e_sRR8fp.js";
import "./models-config.providers.discovery-Ch23kOpG.js";
import "./text-runtime-B9_24WY5.js";
import "./tool-catalog-6VPSKPp9.js";
import "./docker-YL21-mX4.js";
import "./sandbox-C3lR2qEl.js";
import "./common-De-BVwIv.js";
import "./image-ops-BLXW6hKm.js";
import "./thinking-CSfcimRZ.js";
import "./path-alias-guards-CgP-6a1d.js";
import "./sandbox-paths-Bk2ZwMsR.js";
import "./channel-actions-BzT1SAu9.js";
import "./mime-D7Q9o2pi.js";
import "./ssrf-rHmQXFZ4.js";
import "./fetch-guard-BH3MjZoA.js";
import "./provider-web-search-DYdDI2c9.js";
import "./manager-CTkU8M9E.js";
import "./config-schema-Bpwy_blm.js";
import "./channel-plugin-common-COvcxNrG.js";
import "./status-helpers-Dd-hr4j4.js";
import "./conversation-runtime-CqOm9f5F.js";
import "./runtime-whatsapp-boundary-CWceJs-_.js";
import "./pairing-store-ClZkWB4v.js";
import "./json-store-Ju23bywi.js";
import "./secret-file-Ctk8Gt6Y.js";
import "./channel-config-schema-DGuU5OU5.js";
import "./setup-binary-DXdzcWG_.js";
import "./archive-Bk8HKpaY.js";
import "./fs-safe-Dvzkbqib.js";
import "./signal-cli-install-BJB9N18P.js";
import "./setup-wizard-proxy-D0W3ulU0.js";
import "./setup-BYrV6xB7.js";
import "./dm-policy-shared-DSMJu_ZS.js";
import "./hook-runtime-DEbp5SYL.js";
import "./diagnostic-PF4suwLJ.js";
import "./templating-BueU7auS.js";
import "./channel-reply-pipeline-C0iCnI2R.js";
import "./reply-history-BVE7m0XA.js";
import "./commands-registry.data-Dr5wKe-o.js";
import "./commands-registry-C-vLUx3f.js";
import "./frontmatter-DUZeCF_V.js";
import "./env-overrides-DbK0fm-W.js";
import "./skills-CreXENx9.js";
import "./skills-remote-Ctc3qkkj.js";
import "./workspace-dirs-CKrvld61.js";
import "./pairing-token-CT1j20jA.js";
import "./skill-commands-DDonxE3L.js";
import "./level-overrides-DoL5t4pG.js";
import "./config-BkPtspys.js";
import "./routes-CdaHRCET.js";
import "./ssh-tunnel-CnI7n7iv.js";
import "./server-middleware-DcKqo5DB.js";
import "./logging-CSCaAwMS.js";
import "./config-runtime-ST43sCMq.js";
import "./exec-approvals-CVzy3Vfp.js";
import "./webhook-ingress-Chu9Ldhy.js";
import "./system-events-D9mayVHC.js";
import "./ssrf-policy-C4n9geuA.js";
import "./provider-auth-ref-D7iRHU8I.js";
import "./provider-auth-helpers-BeqTsaJO.js";
import "./provider-api-key-auth-C6RcLhw0.js";
import "./runtime-env-CjqWVeEh.js";
import "./pairing-labels-BOG3E444.js";
import "./directory-runtime-Dv9ukfmB.js";
import "./read-only-account-inspect-gnHd6Zsr.js";
import "./web-media-C4M86664.js";
import "./temp-path-BPp7EkAA.js";
import "./media-understanding-BGz4x5yF.js";
import "./web-media-ChNoP5pI.js";
import "./state-paths-kGWIOBBS.js";
import "./llm-task-CVbsbiq5.js";
import "./pi-model-discovery-BpWH93Dy.js";
import "./exec-inline-eval-BXBKoGM8.js";
import "./target-registry-Che-i6sL.js";
import "./external-content-DD7F5dje.js";
import "./brave-DlJT7GTf.js";
import "./duckduckgo-Bkn2wdAA.js";
import "./exa-DpNIw7fA.js";
import "./security-runtime-kOJWyLps.js";
import "./provider-usage-BWl6XLRe.js";
import "./provider-onboard-DSTfYPeq.js";
import "./perplexity-Diz33fcS.js";
import "./stagger-BkeVxPT8.js";
import "./command-secret-targets-BYcKChcy.js";
import "./delivery-queue-gz62qXSE.js";
import "./channel-summary-wFTF6Nzp.js";
import "./session-system-events-CXg2CZLB.js";
import "./tool-policy-match-BNFjIlBI.js";
import "./runtime-DB0HL3pI.js";
import "./channel-status-BsYjzyR-.js";
import "./discord-core-bSS6QgzK.js";
import "./cli-runtime-hF9d0wpB.js";
import "./config-presence-CovoxGn8.js";
import "./query-expansion-CzcRNQqC.js";
import "./search-manager-PSAQEKfx.js";
import "./acp-runtime-BUMdU4AF.js";
import "./telegram-core-BaYrcVUp.js";
import "./audit-CO95QNfX.js";
import "./gateway-runtime-DhcHrysu.js";
import "./connection-auth-COroqkf4.js";
import "./mcp-config-Dp646rw6.js";
import "./provider-auth-login-Bcb4ZSD2.js";
import "./auth-health-BRcIzvG0.js";
import { n as runCommandWithRuntime, t as resolveOptionFromCommand } from "./cli-utils-DwDnCzTo.js";
import "./note-DFCwFXVH.js";
import "./clack-prompter-DKnh22on.js";
import "./table-D77MmWwB.js";
import { C as modelsAliasesRemoveCommand, S as modelsAliasesListCommand, _ as modelsAuthAddCommand, a as modelsListCommand, b as modelsAuthSetupTokenCommand, c as modelsImageFallbacksListCommand, d as modelsFallbacksClearCommand, f as modelsFallbacksListCommand, g as modelsAuthOrderSetCommand, h as modelsAuthOrderGetCommand, i as modelsStatusCommand, l as modelsImageFallbacksRemoveCommand, m as modelsAuthOrderClearCommand, n as modelsSetCommand, o as modelsImageFallbacksAddCommand, p as modelsFallbacksRemoveCommand, r as modelsScanCommand, s as modelsImageFallbacksClearCommand, t as modelsSetImageCommand, u as modelsFallbacksAddCommand, v as modelsAuthLoginCommand, x as modelsAliasesAddCommand, y as modelsAuthPasteTokenCommand } from "./models-C6Rr59E2.js";
import "./provider-auth-choice-helpers-BeLNIaNI.js";
import "./provider-auth-guidance-jLdnZwhv.js";
//#region src/cli/models-cli.ts
function runModelsCommand(action) {
	return runCommandWithRuntime(defaultRuntime, action);
}
function registerModelsCli(program) {
	const models = program.command("models").description("Model discovery, scanning, and configuration").option("--status-json", "Output JSON (alias for `models status --json`)", false).option("--status-plain", "Plain output (alias for `models status --plain`)", false).option("--agent <id>", "Agent id to inspect (overrides OPENCLAW_AGENT_DIR/PI_CODING_AGENT_DIR)").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/models", "docs.openclaw.ai/cli/models")}\n`);
	models.command("list").description("List models (configured by default)").option("--all", "Show full model catalog", false).option("--local", "Filter to local models", false).option("--provider <name>", "Filter by provider").option("--json", "Output JSON", false).option("--plain", "Plain line output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsListCommand(opts, defaultRuntime);
		});
	});
	models.command("status").description("Show configured model state").option("--json", "Output JSON", false).option("--plain", "Plain output", false).option("--check", "Exit non-zero if auth is expiring/expired (1=expired/missing, 2=expiring)", false).option("--probe", "Probe configured provider auth (live)", false).option("--probe-provider <name>", "Only probe a single provider").option("--probe-profile <id>", "Only probe specific auth profile ids (repeat or comma-separated)", (value, previous) => {
		const next = Array.isArray(previous) ? previous : previous ? [previous] : [];
		next.push(value);
		return next;
	}).option("--probe-timeout <ms>", "Per-probe timeout in ms").option("--probe-concurrency <n>", "Concurrent probes").option("--probe-max-tokens <n>", "Probe max tokens (best-effort)").option("--agent <id>", "Agent id to inspect (overrides OPENCLAW_AGENT_DIR/PI_CODING_AGENT_DIR)").action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsStatusCommand({
				json: Boolean(opts.json),
				plain: Boolean(opts.plain),
				check: Boolean(opts.check),
				probe: Boolean(opts.probe),
				probeProvider: opts.probeProvider,
				probeProfile: opts.probeProfile,
				probeTimeout: opts.probeTimeout,
				probeConcurrency: opts.probeConcurrency,
				probeMaxTokens: opts.probeMaxTokens,
				agent
			}, defaultRuntime);
		});
	});
	models.command("set").description("Set the default model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsSetCommand(model, defaultRuntime);
		});
	});
	models.command("set-image").description("Set the image model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsSetImageCommand(model, defaultRuntime);
		});
	});
	const aliases = models.command("aliases").description("Manage model aliases");
	aliases.command("list").description("List model aliases").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAliasesListCommand(opts, defaultRuntime);
		});
	});
	aliases.command("add").description("Add or update a model alias").argument("<alias>", "Alias name").argument("<model>", "Model id or alias").action(async (alias, model) => {
		await runModelsCommand(async () => {
			await modelsAliasesAddCommand(alias, model, defaultRuntime);
		});
	});
	aliases.command("remove").description("Remove a model alias").argument("<alias>", "Alias name").action(async (alias) => {
		await runModelsCommand(async () => {
			await modelsAliasesRemoveCommand(alias, defaultRuntime);
		});
	});
	const fallbacks = models.command("fallbacks").description("Manage model fallback list");
	fallbacks.command("list").description("List fallback models").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsFallbacksListCommand(opts, defaultRuntime);
		});
	});
	fallbacks.command("add").description("Add a fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsFallbacksAddCommand(model, defaultRuntime);
		});
	});
	fallbacks.command("remove").description("Remove a fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsFallbacksRemoveCommand(model, defaultRuntime);
		});
	});
	fallbacks.command("clear").description("Clear all fallback models").action(async () => {
		await runModelsCommand(async () => {
			await modelsFallbacksClearCommand(defaultRuntime);
		});
	});
	const imageFallbacks = models.command("image-fallbacks").description("Manage image model fallback list");
	imageFallbacks.command("list").description("List image fallback models").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksListCommand(opts, defaultRuntime);
		});
	});
	imageFallbacks.command("add").description("Add an image fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksAddCommand(model, defaultRuntime);
		});
	});
	imageFallbacks.command("remove").description("Remove an image fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksRemoveCommand(model, defaultRuntime);
		});
	});
	imageFallbacks.command("clear").description("Clear all image fallback models").action(async () => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksClearCommand(defaultRuntime);
		});
	});
	models.command("scan").description("Scan OpenRouter free models for tools + images").option("--min-params <b>", "Minimum parameter size (billions)").option("--max-age-days <days>", "Skip models older than N days").option("--provider <name>", "Filter by provider prefix").option("--max-candidates <n>", "Max fallback candidates", "6").option("--timeout <ms>", "Per-probe timeout in ms").option("--concurrency <n>", "Probe concurrency").option("--no-probe", "Skip live probes; list free candidates only").option("--yes", "Accept defaults without prompting", false).option("--no-input", "Disable prompts (use defaults)").option("--set-default", "Set agents.defaults.model to the first selection", false).option("--set-image", "Set agents.defaults.imageModel to the first image selection", false).option("--json", "Output JSON", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsScanCommand(opts, defaultRuntime);
		});
	});
	models.action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsStatusCommand({
				json: Boolean(opts?.statusJson),
				plain: Boolean(opts?.statusPlain),
				agent: opts?.agent
			}, defaultRuntime);
		});
	});
	const auth = models.command("auth").description("Manage model auth profiles");
	auth.option("--agent <id>", "Agent id for auth order get/set/clear");
	auth.action(() => {
		auth.help();
	});
	auth.command("add").description("Interactive auth helper (setup-token or paste token)").action(async () => {
		await runModelsCommand(async () => {
			await modelsAuthAddCommand({}, defaultRuntime);
		});
	});
	auth.command("login").description("Run a provider plugin auth flow (OAuth/API key)").option("--provider <id>", "Provider id registered by a plugin").option("--method <id>", "Provider auth method id").option("--set-default", "Apply the provider's default model recommendation", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthLoginCommand({
				provider: opts.provider,
				method: opts.method,
				setDefault: Boolean(opts.setDefault)
			}, defaultRuntime);
		});
	});
	auth.command("setup-token").description("Run a provider CLI to create/sync a token (TTY required)").option("--provider <name>", "Provider id (default: anthropic)").option("--yes", "Skip confirmation", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthSetupTokenCommand({
				provider: opts.provider,
				yes: Boolean(opts.yes)
			}, defaultRuntime);
		});
	});
	auth.command("paste-token").description("Paste a token into auth-profiles.json and update config").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--profile-id <id>", "Auth profile id (default: <provider>:manual)").option("--expires-in <duration>", "Optional expiry duration (e.g. 365d, 12h). Stored as absolute expiresAt.").action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthPasteTokenCommand({
				provider: opts.provider,
				profileId: opts.profileId,
				expiresIn: opts.expiresIn
			}, defaultRuntime);
		});
	});
	auth.command("login-github-copilot").description("Login to GitHub Copilot via GitHub device flow (TTY required)").option("--yes", "Overwrite existing profile without prompting", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthLoginCommand({
				provider: "github-copilot",
				method: "device",
				yes: Boolean(opts.yes)
			}, defaultRuntime);
		});
	});
	const order = auth.command("order").description("Manage per-agent auth profile order overrides");
	order.command("get").description("Show per-agent auth order override (from auth-profiles.json)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").option("--json", "Output JSON", false).action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderGetCommand({
				provider: opts.provider,
				agent,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	order.command("set").description("Set per-agent auth order override (locks rotation to this list)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").argument("<profileIds...>", "Auth profile ids (e.g. anthropic:default)").action(async (profileIds, opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderSetCommand({
				provider: opts.provider,
				agent,
				order: profileIds
			}, defaultRuntime);
		});
	});
	order.command("clear").description("Clear per-agent auth order override (fall back to config/round-robin)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderClearCommand({
				provider: opts.provider,
				agent
			}, defaultRuntime);
		});
	});
}
//#endregion
export { registerModelsCli };
