// openclaw/dist/oauth.runtime-BUJG3ptH.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import "./theme-CdOoMzRk.js";
import "./globals-DBUMOBZ8.js";
import "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import "./utils-CS0Ikux6.js";
import "./links-8xRhWBQL.js";
import "./paths-BFl2-hCf.js";
import "./auth-profiles-CVFpgeL-.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import "./agent-scope-DoT9OqaV.js";
import "./model-selection-B6ao45a4.js";
import "./io-cPs4dU7X.js";
import "./host-env-security-DEKL50zA.js";
import "./shell-env-BKuVS72k.js";
import "./safe-text-DIDDfQyI.js";
import "./version-BpHNkJed.js";
import "./env-substitution-B8NFl2Jd.js";
import "./includes-CCK6fRRs.js";
import "./zod-schema.providers-core-DLPfih2y.js";
import "./legacy-web-search-CTvir-Bl.js";
import "./registry-xyHjVLxh.js";
import "./config-state-Dx9-tLPS.js";
import "./min-host-version-xbc6BJ_K.js";
import "./manifest-registry-D5E7Gxgl.js";
import "./runtime-guard-QGt7fm0l.js";
import "./avatar-policy-DVf9B9eu.js";
import "./ip-DIgtRRTW.js";
import "./zod-schema.agent-runtime-C659GPl8.js";
import "./zod-schema.core-yLTNC4-K.js";
import "./config-CB4aYWqd.js";
import "./file-lock-DoKDW8jx.js";
import "./audit-fs-CEN00XrG.js";
import "./resolve-uXRpJb-M.js";
import "./profiles-Cyg58FCO.js";
import "./tailscale-CSUwCuE9.js";
import "./tailnet-BVzEE6AW.js";
import "./net-Dk658jWW.js";
import "./auth-C_pTuZtn.js";
import "./credentials-CO55Yx_u.js";
import "./message-channel-DUrzQUcI.js";
import "./store-D_eNuCKK.js";
import "./runtime-C8dQugND.js";
import "./plugins-DGdgUNcN.js";
import "./sessions-C8WesgCl.js";
import "./paths-BPfguEtH.js";
import "./session-write-lock-bE2vQvvO.js";
import "./method-scopes-BWG4Q18M.js";
import "./call-VZCb020X.js";
import { n as isWSL2Sync } from "./wsl-DoeW56DL.js";
import "./prompt-style-MhFLSlua.js";
import "./ports-lsof-C3NOmuKA.js";
import "./restart-stale-pids-CQbYk99d.js";
import "./ports-CQIuVpXl.js";
import "./logging-C2L37N2X.js";
import "./commands-CrPoYX9r.js";
import "./issue-format-CP-gqjZB.js";
import "./identity-D8H54Ni5.js";
import "./heartbeat-FZaB85hC.js";
import "./pi-embedded-CzQCqSlH.js";
import "./internal-hooks-D4lZfNM5.js";
import "./multimodal-De3qv_72.js";
import "./memory-search-DRm4OVlH.js";
import "./provider-catalog-7bUKai5_.js";
import "./secret-input-BNBMqe2K.js";
import "./bindings-RcwGGOd3.js";
import "./resolve-route-DbP_XX4E.js";
import "./routing-B37UwRwq.js";
import "./identity-file-CCks_qJo.js";
import "./outbound-runtime-oDjCAP7B.js";
import "./provider-env-vars-Bhzl_gIs.js";
import "./provider-auth-input-DxjPWsSV.js";
import "./provider-model-minimax-XdwKeUOE.js";
import "./provider-models-DfgBoVpf.js";
import "./anthropic-vertex-provider-e_sRR8fp.js";
import "./models-config.providers.discovery-Ch23kOpG.js";
import "./text-runtime-B9_24WY5.js";
import "./tool-catalog-6VPSKPp9.js";
import "./docker-YL21-mX4.js";
import "./sandbox-C3lR2qEl.js";
import "./common-De-BVwIv.js";
import "./image-ops-BLXW6hKm.js";
import "./thinking-CSfcimRZ.js";
import "./path-alias-guards-CgP-6a1d.js";
import "./sandbox-paths-Bk2ZwMsR.js";
import "./channel-actions-BzT1SAu9.js";
import "./mime-D7Q9o2pi.js";
import "./ssrf-rHmQXFZ4.js";
import { n as fetchWithSsrFGuard } from "./fetch-guard-BH3MjZoA.js";
import "./provider-web-search-DYdDI2c9.js";
import "./manager-CTkU8M9E.js";
import "./config-schema-Bpwy_blm.js";
import "./channel-plugin-common-COvcxNrG.js";
import "./status-helpers-Dd-hr4j4.js";
import "./conversation-runtime-CqOm9f5F.js";
import "./runtime-whatsapp-boundary-CWceJs-_.js";
import "./pairing-store-ClZkWB4v.js";
import "./json-store-Ju23bywi.js";
import "./secret-file-Ctk8Gt6Y.js";
import "./channel-config-schema-DGuU5OU5.js";
import "./setup-binary-DXdzcWG_.js";
import "./archive-Bk8HKpaY.js";
import "./fs-safe-Dvzkbqib.js";
import "./signal-cli-install-BJB9N18P.js";
import "./setup-wizard-proxy-D0W3ulU0.js";
import "./setup-BYrV6xB7.js";
import "./dm-policy-shared-DSMJu_ZS.js";
import "./hook-runtime-DEbp5SYL.js";
import "./diagnostic-PF4suwLJ.js";
import "./templating-BueU7auS.js";
import "./channel-reply-pipeline-C0iCnI2R.js";
import "./reply-history-BVE7m0XA.js";
import "./commands-registry.data-Dr5wKe-o.js";
import "./commands-registry-C-vLUx3f.js";
import "./frontmatter-DUZeCF_V.js";
import "./env-overrides-DbK0fm-W.js";
import "./skills-CreXENx9.js";
import "./skills-remote-Ctc3qkkj.js";
import "./workspace-dirs-CKrvld61.js";
import "./pairing-token-CT1j20jA.js";
import "./skill-commands-DDonxE3L.js";
import "./level-overrides-DoL5t4pG.js";
import "./config-BkPtspys.js";
import "./routes-CdaHRCET.js";
import "./ssh-tunnel-CnI7n7iv.js";
import "./server-middleware-DcKqo5DB.js";
import "./logging-CSCaAwMS.js";
import "./config-runtime-ST43sCMq.js";
import "./exec-approvals-CVzy3Vfp.js";
import "./webhook-ingress-Chu9Ldhy.js";
import "./system-events-D9mayVHC.js";
import "./ssrf-policy-C4n9geuA.js";
import "./provider-auth-ref-D7iRHU8I.js";
import "./provider-auth-helpers-BeqTsaJO.js";
import "./provider-api-key-auth-C6RcLhw0.js";
import "./runtime-env-CjqWVeEh.js";
import "./pairing-labels-BOG3E444.js";
import "./directory-runtime-Dv9ukfmB.js";
import "./read-only-account-inspect-gnHd6Zsr.js";
import "./web-media-C4M86664.js";
import "./temp-path-BPp7EkAA.js";
import "./media-understanding-BGz4x5yF.js";
import "./web-media-ChNoP5pI.js";
import "./state-paths-kGWIOBBS.js";
import "./llm-task-CVbsbiq5.js";
import "./pi-model-discovery-BpWH93Dy.js";
import "./exec-inline-eval-BXBKoGM8.js";
import "./target-registry-Che-i6sL.js";
import "./external-content-DD7F5dje.js";
import "./brave-DlJT7GTf.js";
import "./duckduckgo-Bkn2wdAA.js";
import "./exa-DpNIw7fA.js";
import "./security-runtime-kOJWyLps.js";
import "./provider-usage-BWl6XLRe.js";
import "./provider-onboard-DSTfYPeq.js";
import "./perplexity-Diz33fcS.js";
import "./stagger-BkeVxPT8.js";
import "./command-secret-targets-BYcKChcy.js";
import "./delivery-queue-gz62qXSE.js";
import "./channel-summary-wFTF6Nzp.js";
import "./session-system-events-CXg2CZLB.js";
import "./tool-policy-match-BNFjIlBI.js";
import "./runtime-DB0HL3pI.js";
import "./channel-status-BsYjzyR-.js";
import "./discord-core-bSS6QgzK.js";
import "./cli-runtime-hF9d0wpB.js";
import "./config-presence-CovoxGn8.js";
import "./query-expansion-CzcRNQqC.js";
import "./search-manager-PSAQEKfx.js";
import "./acp-runtime-BUMdU4AF.js";
import "./telegram-core-BaYrcVUp.js";
import "./audit-CO95QNfX.js";
import "./gateway-runtime-DhcHrysu.js";
import "./connection-auth-COroqkf4.js";
import "./mcp-config-Dp646rw6.js";
import { existsSync, readFileSync, readdirSync, realpathSync } from "node:fs";
import { delimiter, dirname, join } from "node:path";
import { createHash, randomBytes } from "node:crypto";
import { createServer } from "node:http";
//#region extensions/google/oauth.shared.ts
const CLIENT_ID_KEYS = ["OPENCLAW_GEMINI_OAUTH_CLIENT_ID", "GEMINI_CLI_OAUTH_CLIENT_ID"];
const CLIENT_SECRET_KEYS = ["OPENCLAW_GEMINI_OAUTH_CLIENT_SECRET", "GEMINI_CLI_OAUTH_CLIENT_SECRET"];
const REDIRECT_URI = "http://localhost:8085/oauth2callback";
const AUTH_URL = "https://accounts.google.com/o/oauth2/v2/auth";
const TOKEN_URL = "https://oauth2.googleapis.com/token";
const USERINFO_URL = "https://www.googleapis.com/oauth2/v1/userinfo?alt=json";
const CODE_ASSIST_ENDPOINT_PROD = "https://cloudcode-pa.googleapis.com";
const LOAD_CODE_ASSIST_ENDPOINTS = [
	CODE_ASSIST_ENDPOINT_PROD,
	"https://daily-cloudcode-pa.sandbox.googleapis.com",
	"https://autopush-cloudcode-pa.sandbox.googleapis.com"
];
const DEFAULT_FETCH_TIMEOUT_MS = 1e4;
const SCOPES = [
	"https://www.googleapis.com/auth/cloud-platform",
	"https://www.googleapis.com/auth/userinfo.email",
	"https://www.googleapis.com/auth/userinfo.profile"
];
const TIER_LEGACY = "legacy-tier";
const TIER_STANDARD = "standard-tier";
let credentialFs = {
	existsSync,
	readFileSync,
	realpathSync,
	readdirSync
};
function resolveEnv(keys) {
	for (const key of keys) {
		const value = process.env[key]?.trim();
		if (value) return value;
	}
}
let cachedGeminiCliCredentials = null;
function extractGeminiCliCredentials() {
	if (cachedGeminiCliCredentials) return cachedGeminiCliCredentials;
	try {
		const geminiPath = findInPath("gemini");
		if (!geminiPath) return null;
		const geminiCliDirs = resolveGeminiCliDirs(geminiPath, credentialFs.realpathSync(geminiPath));
		let content = null;
		for (const geminiCliDir of geminiCliDirs) {
			const searchPaths = [join(geminiCliDir, "node_modules", "@google", "gemini-cli-core", "dist", "src", "code_assist", "oauth2.js"), join(geminiCliDir, "node_modules", "@google", "gemini-cli-core", "dist", "code_assist", "oauth2.js")];
			for (const path of searchPaths) if (credentialFs.existsSync(path)) {
				content = credentialFs.readFileSync(path, "utf8");
				break;
			}
			if (content) break;
			const found = findFile(geminiCliDir, "oauth2.js", 10);
			if (found) {
				content = credentialFs.readFileSync(found, "utf8");
				break;
			}
		}
		if (!content) return null;
		const idMatch = content.match(/(\d+-[a-z0-9]+\.apps\.googleusercontent\.com)/);
		const secretMatch = content.match(/(GOCSPX-[A-Za-z0-9_-]+)/);
		if (idMatch && secretMatch) {
			cachedGeminiCliCredentials = {
				clientId: idMatch[1],
				clientSecret: secretMatch[1]
			};
			return cachedGeminiCliCredentials;
		}
	} catch {}
	return null;
}
function resolveGeminiCliDirs(geminiPath, resolvedPath) {
	const binDir = dirname(geminiPath);
	const candidates = [
		dirname(dirname(resolvedPath)),
		join(dirname(resolvedPath), "node_modules", "@google", "gemini-cli"),
		join(binDir, "node_modules", "@google", "gemini-cli"),
		join(dirname(binDir), "node_modules", "@google", "gemini-cli"),
		join(dirname(binDir), "lib", "node_modules", "@google", "gemini-cli")
	];
	const deduped = [];
	const seen = /* @__PURE__ */ new Set();
	for (const candidate of candidates) {
		const key = process.platform === "win32" ? candidate.replace(/\\/g, "/").toLowerCase() : candidate;
		if (seen.has(key)) continue;
		seen.add(key);
		deduped.push(candidate);
	}
	return deduped;
}
function findInPath(name) {
	const exts = process.platform === "win32" ? [
		".cmd",
		".bat",
		".exe",
		""
	] : [""];
	for (const dir of (process.env.PATH ?? "").split(delimiter)) for (const ext of exts) {
		const path = join(dir, name + ext);
		if (credentialFs.existsSync(path)) return path;
	}
	return null;
}
function findFile(dir, name, depth) {
	if (depth <= 0) return null;
	try {
		for (const entry of credentialFs.readdirSync(dir, { withFileTypes: true })) {
			const path = join(dir, entry.name);
			if (entry.isFile() && entry.name === name) return path;
			if (entry.isDirectory() && !entry.name.startsWith(".")) {
				const found = findFile(path, name, depth - 1);
				if (found) return found;
			}
		}
	} catch {}
	return null;
}
function resolveOAuthClientConfig() {
	const envClientId = resolveEnv(CLIENT_ID_KEYS);
	const envClientSecret = resolveEnv(CLIENT_SECRET_KEYS);
	if (envClientId) return {
		clientId: envClientId,
		clientSecret: envClientSecret
	};
	const extracted = extractGeminiCliCredentials();
	if (extracted) return extracted;
	throw new Error("Gemini CLI not found. Install it first: brew install gemini-cli (or npm install -g @google/gemini-cli), or set GEMINI_CLI_OAUTH_CLIENT_ID.");
}
//#endregion
//#region extensions/google/oauth.flow.ts
function shouldUseManualOAuthFlow(isRemote) {
	return isRemote || isWSL2Sync();
}
function generatePkce() {
	const verifier = randomBytes(32).toString("hex");
	return {
		verifier,
		challenge: createHash("sha256").update(verifier).digest("base64url")
	};
}
function buildAuthUrl(challenge, verifier) {
	const { clientId } = resolveOAuthClientConfig();
	return `${AUTH_URL}?${new URLSearchParams({
		client_id: clientId,
		response_type: "code",
		redirect_uri: REDIRECT_URI,
		scope: SCOPES.join(" "),
		code_challenge: challenge,
		code_challenge_method: "S256",
		state: verifier,
		access_type: "offline",
		prompt: "consent"
	}).toString()}`;
}
function parseCallbackInput(input, expectedState) {
	const trimmed = input.trim();
	if (!trimmed) return { error: "No input provided" };
	try {
		const url = new URL(trimmed);
		const code = url.searchParams.get("code");
		const state = url.searchParams.get("state") ?? expectedState;
		if (!code) return { error: "Missing 'code' parameter in URL" };
		if (!state) return { error: "Missing 'state' parameter. Paste the full URL." };
		return {
			code,
			state
		};
	} catch {
		if (!expectedState) return { error: "Paste the full redirect URL, not just the code." };
		return {
			code: trimmed,
			state: expectedState
		};
	}
}
async function waitForLocalCallback(params) {
	const port = 8085;
	const hostname = "localhost";
	const expectedPath = "/oauth2callback";
	return new Promise((resolve, reject) => {
		let timeout = null;
		const server = createServer((req, res) => {
			try {
				const requestUrl = new URL(req.url ?? "/", `http://${hostname}:${port}`);
				if (requestUrl.pathname !== expectedPath) {
					res.statusCode = 404;
					res.setHeader("Content-Type", "text/plain");
					res.end("Not found");
					return;
				}
				const error = requestUrl.searchParams.get("error");
				const code = requestUrl.searchParams.get("code")?.trim();
				const state = requestUrl.searchParams.get("state")?.trim();
				if (error) {
					res.statusCode = 400;
					res.setHeader("Content-Type", "text/plain");
					res.end(`Authentication failed: ${error}`);
					finish(/* @__PURE__ */ new Error(`OAuth error: ${error}`));
					return;
				}
				if (!code || !state) {
					res.statusCode = 400;
					res.setHeader("Content-Type", "text/plain");
					res.end("Missing code or state");
					finish(/* @__PURE__ */ new Error("Missing OAuth code or state"));
					return;
				}
				if (state !== params.expectedState) {
					res.statusCode = 400;
					res.setHeader("Content-Type", "text/plain");
					res.end("Invalid state");
					finish(/* @__PURE__ */ new Error("OAuth state mismatch"));
					return;
				}
				res.statusCode = 200;
				res.setHeader("Content-Type", "text/html; charset=utf-8");
				res.end("<!doctype html><html><head><meta charset='utf-8'/></head><body><h2>Gemini CLI OAuth complete</h2><p>You can close this window and return to OpenClaw.</p></body></html>");
				finish(void 0, {
					code,
					state
				});
			} catch (err) {
				finish(err instanceof Error ? err : /* @__PURE__ */ new Error("OAuth callback failed"));
			}
		});
		const finish = (err, result) => {
			if (timeout) clearTimeout(timeout);
			try {
				server.close();
			} catch {}
			if (err) reject(err);
			else if (result) resolve(result);
		};
		server.once("error", (err) => {
			finish(err instanceof Error ? err : /* @__PURE__ */ new Error("OAuth callback server error"));
		});
		server.listen(port, hostname, () => {
			params.onProgress?.(`Waiting for OAuth callback on ${REDIRECT_URI}…`);
		});
		timeout = setTimeout(() => {
			finish(/* @__PURE__ */ new Error("OAuth callback timeout"));
		}, params.timeoutMs);
	});
}
//#endregion
//#region extensions/google/oauth.http.ts
async function fetchWithTimeout(url, init, timeoutMs = DEFAULT_FETCH_TIMEOUT_MS) {
	const { response, release } = await fetchWithSsrFGuard({
		url,
		init,
		timeoutMs
	});
	try {
		const body = await response.arrayBuffer();
		return new Response(body, {
			status: response.status,
			statusText: response.statusText,
			headers: response.headers
		});
	} finally {
		await release();
	}
}
//#endregion
//#region extensions/google/oauth.project.ts
function resolvePlatform() {
	if (process.platform === "win32") return "WINDOWS";
	if (process.platform === "darwin") return "MACOS";
	return "PLATFORM_UNSPECIFIED";
}
async function getUserEmail(accessToken) {
	try {
		const response = await fetchWithTimeout(USERINFO_URL, { headers: { Authorization: `Bearer ${accessToken}` } });
		if (response.ok) return (await response.json()).email;
	} catch {}
}
function isVpcScAffected(payload) {
	if (!payload || typeof payload !== "object") return false;
	const error = payload.error;
	if (!error || typeof error !== "object") return false;
	const details = error.details;
	if (!Array.isArray(details)) return false;
	return details.some((item) => typeof item === "object" && item && item.reason === "SECURITY_POLICY_VIOLATED");
}
function getDefaultTier(allowedTiers) {
	if (!allowedTiers?.length) return { id: TIER_LEGACY };
	return allowedTiers.find((tier) => tier.isDefault) ?? { id: "legacy-tier" };
}
async function pollOperation(endpoint, operationName, headers) {
	for (let attempt = 0; attempt < 24; attempt += 1) {
		await new Promise((resolve) => setTimeout(resolve, 5e3));
		const response = await fetchWithTimeout(`${endpoint}/v1internal/${operationName}`, { headers });
		if (!response.ok) continue;
		const data = await response.json();
		if (data.done) return data;
	}
	throw new Error("Operation polling timeout");
}
async function resolveGoogleOAuthIdentity(accessToken) {
	return {
		email: await getUserEmail(accessToken),
		projectId: await discoverProject(accessToken)
	};
}
async function discoverProject(accessToken) {
	const envProject = process.env.GOOGLE_CLOUD_PROJECT || process.env.GOOGLE_CLOUD_PROJECT_ID;
	const metadata = {
		ideType: "ANTIGRAVITY",
		platform: resolvePlatform(),
		pluginType: "GEMINI"
	};
	const headers = {
		Authorization: `Bearer ${accessToken}`,
		"Content-Type": "application/json",
		"User-Agent": "google-api-nodejs-client/9.15.1",
		"X-Goog-Api-Client": `gl-node/${process.versions.node}`,
		"Client-Metadata": JSON.stringify(metadata)
	};
	const loadBody = {
		...envProject ? { cloudaicompanionProject: envProject } : {},
		metadata: {
			...metadata,
			...envProject ? { duetProject: envProject } : {}
		}
	};
	let data = {};
	let activeEndpoint = CODE_ASSIST_ENDPOINT_PROD;
	let loadError;
	for (const endpoint of LOAD_CODE_ASSIST_ENDPOINTS) try {
		const response = await fetchWithTimeout(`${endpoint}/v1internal:loadCodeAssist`, {
			method: "POST",
			headers,
			body: JSON.stringify(loadBody)
		});
		if (!response.ok) {
			if (isVpcScAffected(await response.json().catch(() => null))) {
				data = { currentTier: { id: TIER_STANDARD } };
				activeEndpoint = endpoint;
				loadError = void 0;
				break;
			}
			loadError = /* @__PURE__ */ new Error(`loadCodeAssist failed: ${response.status} ${response.statusText}`);
			continue;
		}
		data = await response.json();
		activeEndpoint = endpoint;
		loadError = void 0;
		break;
	} catch (err) {
		loadError = err instanceof Error ? err : new Error("loadCodeAssist failed", { cause: err });
	}
	if (!(Boolean(data.currentTier) || Boolean(data.cloudaicompanionProject) || Boolean(data.allowedTiers?.length)) && loadError) {
		if (envProject) return envProject;
		throw loadError;
	}
	if (data.currentTier) {
		const project = data.cloudaicompanionProject;
		if (typeof project === "string" && project) return project;
		if (typeof project === "object" && project?.id) return project.id;
		if (envProject) return envProject;
		throw new Error("This account requires GOOGLE_CLOUD_PROJECT or GOOGLE_CLOUD_PROJECT_ID to be set.");
	}
	const tierId = getDefaultTier(data.allowedTiers)?.id || "free-tier";
	if (tierId !== "free-tier" && !envProject) throw new Error("This account requires GOOGLE_CLOUD_PROJECT or GOOGLE_CLOUD_PROJECT_ID to be set.");
	const onboardBody = {
		tierId,
		metadata: { ...metadata }
	};
	if (tierId !== "free-tier" && envProject) {
		onboardBody.cloudaicompanionProject = envProject;
		onboardBody.metadata.duetProject = envProject;
	}
	const onboardResponse = await fetchWithTimeout(`${activeEndpoint}/v1internal:onboardUser`, {
		method: "POST",
		headers,
		body: JSON.stringify(onboardBody)
	});
	if (!onboardResponse.ok) throw new Error(`onboardUser failed: ${onboardResponse.status} ${onboardResponse.statusText}`);
	let lro = await onboardResponse.json();
	if (!lro.done && lro.name) lro = await pollOperation(activeEndpoint, lro.name, headers);
	const projectId = lro.response?.cloudaicompanionProject?.id;
	if (projectId) return projectId;
	if (envProject) return envProject;
	throw new Error("Could not discover or provision a Google Cloud project. Set GOOGLE_CLOUD_PROJECT or GOOGLE_CLOUD_PROJECT_ID.");
}
//#endregion
//#region extensions/google/oauth.token.ts
async function exchangeCodeForTokens(code, verifier) {
	const { clientId, clientSecret } = resolveOAuthClientConfig();
	const body = new URLSearchParams({
		client_id: clientId,
		code,
		grant_type: "authorization_code",
		redirect_uri: REDIRECT_URI,
		code_verifier: verifier
	});
	if (clientSecret) body.set("client_secret", clientSecret);
	const response = await fetchWithTimeout(TOKEN_URL, {
		method: "POST",
		headers: {
			"Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
			Accept: "*/*",
			"User-Agent": "google-api-nodejs-client/9.15.1"
		},
		body
	});
	if (!response.ok) {
		const errorText = await response.text();
		throw new Error(`Token exchange failed: ${errorText}`);
	}
	const data = await response.json();
	if (!data.refresh_token) throw new Error("No refresh token received. Please try again.");
	const identity = await resolveGoogleOAuthIdentity(data.access_token);
	const expiresAt = Date.now() + data.expires_in * 1e3 - 300 * 1e3;
	return {
		refresh: data.refresh_token,
		access: data.access_token,
		expires: expiresAt,
		projectId: identity.projectId,
		email: identity.email
	};
}
//#endregion
//#region extensions/google/oauth.ts
async function loginGeminiCliOAuth(ctx) {
	const needsManual = shouldUseManualOAuthFlow(ctx.isRemote);
	await ctx.note(needsManual ? [
		"You are running in a remote/VPS environment.",
		"A URL will be shown for you to open in your LOCAL browser.",
		"After signing in, copy the redirect URL and paste it back here."
	].join("\n") : [
		"Browser will open for Google authentication.",
		"Sign in with your Google account for Gemini CLI access.",
		"The callback will be captured automatically on localhost:8085."
	].join("\n"), "Gemini CLI OAuth");
	const { verifier, challenge } = generatePkce();
	const authUrl = buildAuthUrl(challenge, verifier);
	if (needsManual) {
		ctx.progress.update("OAuth URL ready");
		ctx.log(`\nOpen this URL in your LOCAL browser:\n\n${authUrl}\n`);
		ctx.progress.update("Waiting for you to paste the callback URL...");
		const parsed = parseCallbackInput(await ctx.prompt("Paste the redirect URL here: "), verifier);
		if ("error" in parsed) throw new Error(parsed.error);
		if (parsed.state !== verifier) throw new Error("OAuth state mismatch - please try again");
		ctx.progress.update("Exchanging authorization code for tokens...");
		return exchangeCodeForTokens(parsed.code, verifier);
	}
	ctx.progress.update("Complete sign-in in browser...");
	try {
		await ctx.openUrl(authUrl);
	} catch {
		ctx.log(`\nOpen this URL in your browser:\n\n${authUrl}\n`);
	}
	try {
		const { code } = await waitForLocalCallback({
			expectedState: verifier,
			timeoutMs: 300 * 1e3,
			onProgress: (msg) => ctx.progress.update(msg)
		});
		ctx.progress.update("Exchanging authorization code for tokens...");
		return await exchangeCodeForTokens(code, verifier);
	} catch (err) {
		if (err instanceof Error && (err.message.includes("EADDRINUSE") || err.message.includes("port") || err.message.includes("listen"))) {
			ctx.progress.update("Local callback server failed. Switching to manual mode...");
			ctx.log(`\nOpen this URL in your LOCAL browser:\n\n${authUrl}\n`);
			const parsed = parseCallbackInput(await ctx.prompt("Paste the redirect URL here: "), verifier);
			if ("error" in parsed) throw new Error(parsed.error, { cause: err });
			if (parsed.state !== verifier) throw new Error("OAuth state mismatch - please try again", { cause: err });
			ctx.progress.update("Exchanging authorization code for tokens...");
			return exchangeCodeForTokens(parsed.code, verifier);
		}
		throw err;
	}
}
//#endregion
export { loginGeminiCliOAuth };
