// openclaw/dist/register.agent-CK8oagtQ.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import { r as theme } from "./theme-CdOoMzRk.js";
import { s as setVerbose } from "./globals-DBUMOBZ8.js";
import { h as writeRuntimeJson, m as defaultRuntime } from "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import "./utils-CS0Ikux6.js";
import { t as formatDocsLink } from "./links-8xRhWBQL.js";
import { t as hasExplicitOptions } from "./command-options-5coRiipK.js";
import "./paths-BFl2-hCf.js";
import "./auth-profiles-CVFpgeL-.js";
import { c as normalizeAgentId } from "./session-key-DAhnzjyr.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import { r as listAgentIds } from "./agent-scope-DoT9OqaV.js";
import "./model-selection-B6ao45a4.js";
import { s as loadConfig } from "./io-cPs4dU7X.js";
import "./host-env-security-DEKL50zA.js";
import "./shell-env-BKuVS72k.js";
import "./safe-text-DIDDfQyI.js";
import "./version-BpHNkJed.js";
import "./env-substitution-B8NFl2Jd.js";
import "./includes-CCK6fRRs.js";
import "./zod-schema.providers-core-DLPfih2y.js";
import "./legacy-web-search-CTvir-Bl.js";
import "./registry-xyHjVLxh.js";
import "./config-state-Dx9-tLPS.js";
import "./min-host-version-xbc6BJ_K.js";
import "./manifest-registry-D5E7Gxgl.js";
import "./runtime-guard-QGt7fm0l.js";
import "./avatar-policy-DVf9B9eu.js";
import "./ip-DIgtRRTW.js";
import "./zod-schema.agent-runtime-C659GPl8.js";
import "./zod-schema.core-yLTNC4-K.js";
import "./config-CB4aYWqd.js";
import "./file-lock-DoKDW8jx.js";
import "./audit-fs-CEN00XrG.js";
import "./resolve-uXRpJb-M.js";
import "./profiles-Cyg58FCO.js";
import { t as formatCliCommand } from "./command-format-DRptdHvm.js";
import "./tailscale-CSUwCuE9.js";
import "./tailnet-BVzEE6AW.js";
import "./net-Dk658jWW.js";
import "./auth-C_pTuZtn.js";
import "./credentials-CO55Yx_u.js";
import { h as GATEWAY_CLIENT_NAMES, l as normalizeMessageChannel, m as GATEWAY_CLIENT_MODES } from "./message-channel-DUrzQUcI.js";
import "./store-D_eNuCKK.js";
import "./runtime-C8dQugND.js";
import "./plugins-DGdgUNcN.js";
import "./sessions-C8WesgCl.js";
import "./paths-BPfguEtH.js";
import "./session-write-lock-bE2vQvvO.js";
import "./method-scopes-BWG4Q18M.js";
import { n as callGateway, s as randomIdempotencyKey } from "./call-VZCb020X.js";
import "./control-ui-shared-SxKiMaO4.js";
import "./onboard-helpers-Cks5EPpw.js";
import "./prompt-style-MhFLSlua.js";
import "./ports-lsof-C3NOmuKA.js";
import "./restart-stale-pids-CQbYk99d.js";
import "./ports-CQIuVpXl.js";
import "./logging-C2L37N2X.js";
import "./commands-CrPoYX9r.js";
import "./issue-format-CP-gqjZB.js";
import { n as withProgress } from "./progress-CWxCPcAp.js";
import { t as formatHelpExamples } from "./help-format-BW49iAcY.js";
import "./identity-D8H54Ni5.js";
import "./heartbeat-FZaB85hC.js";
import { Ti as createDefaultDeps, _i as resolveSessionKeyForRequest, hi as agentCommand } from "./pi-embedded-CzQCqSlH.js";
import "./internal-hooks-D4lZfNM5.js";
import "./multimodal-De3qv_72.js";
import "./memory-search-DRm4OVlH.js";
import "./provider-catalog-7bUKai5_.js";
import "./secret-input-BNBMqe2K.js";
import "./bindings-RcwGGOd3.js";
import "./resolve-route-DbP_XX4E.js";
import "./routing-B37UwRwq.js";
import "./identity-file-CCks_qJo.js";
import "./outbound-runtime-oDjCAP7B.js";
import "./provider-env-vars-Bhzl_gIs.js";
import "./provider-auth-input-DxjPWsSV.js";
import "./provider-model-minimax-XdwKeUOE.js";
import "./provider-models-DfgBoVpf.js";
import "./anthropic-vertex-provider-e_sRR8fp.js";
import "./models-config.providers.discovery-Ch23kOpG.js";
import "./text-runtime-B9_24WY5.js";
import "./tool-catalog-6VPSKPp9.js";
import "./docker-YL21-mX4.js";
import "./sandbox-C3lR2qEl.js";
import "./common-De-BVwIv.js";
import "./image-ops-BLXW6hKm.js";
import "./thinking-CSfcimRZ.js";
import "./path-alias-guards-CgP-6a1d.js";
import "./sandbox-paths-Bk2ZwMsR.js";
import "./channel-actions-BzT1SAu9.js";
import "./mime-D7Q9o2pi.js";
import "./ssrf-rHmQXFZ4.js";
import "./fetch-guard-BH3MjZoA.js";
import "./provider-web-search-DYdDI2c9.js";
import "./manager-CTkU8M9E.js";
import "./config-schema-Bpwy_blm.js";
import "./channel-plugin-common-COvcxNrG.js";
import "./status-helpers-Dd-hr4j4.js";
import "./conversation-runtime-CqOm9f5F.js";
import "./runtime-whatsapp-boundary-CWceJs-_.js";
import "./pairing-store-ClZkWB4v.js";
import "./json-store-Ju23bywi.js";
import "./secret-file-Ctk8Gt6Y.js";
import "./channel-config-schema-DGuU5OU5.js";
import "./setup-binary-DXdzcWG_.js";
import "./archive-Bk8HKpaY.js";
import "./fs-safe-Dvzkbqib.js";
import "./signal-cli-install-BJB9N18P.js";
import "./setup-wizard-proxy-D0W3ulU0.js";
import "./setup-BYrV6xB7.js";
import "./dm-policy-shared-DSMJu_ZS.js";
import { p as resolveSendableOutboundReplyParts } from "./reply-payload-MXtGsVoh.js";
import "./hook-runtime-DEbp5SYL.js";
import "./diagnostic-PF4suwLJ.js";
import "./templating-BueU7auS.js";
import "./channel-reply-pipeline-C0iCnI2R.js";
import "./reply-history-BVE7m0XA.js";
import "./commands-registry.data-Dr5wKe-o.js";
import "./commands-registry-C-vLUx3f.js";
import "./frontmatter-DUZeCF_V.js";
import "./env-overrides-DbK0fm-W.js";
import "./skills-CreXENx9.js";
import "./skills-remote-Ctc3qkkj.js";
import "./workspace-dirs-CKrvld61.js";
import "./pairing-token-CT1j20jA.js";
import "./skill-commands-DDonxE3L.js";
import "./level-overrides-DoL5t4pG.js";
import "./config-BkPtspys.js";
import "./routes-CdaHRCET.js";
import "./ssh-tunnel-CnI7n7iv.js";
import "./server-middleware-DcKqo5DB.js";
import "./logging-CSCaAwMS.js";
import "./config-runtime-ST43sCMq.js";
import "./exec-approvals-CVzy3Vfp.js";
import "./webhook-ingress-Chu9Ldhy.js";
import "./system-events-D9mayVHC.js";
import "./ssrf-policy-C4n9geuA.js";
import "./provider-auth-ref-D7iRHU8I.js";
import "./provider-auth-helpers-BeqTsaJO.js";
import "./provider-api-key-auth-C6RcLhw0.js";
import "./runtime-env-CjqWVeEh.js";
import "./pairing-labels-BOG3E444.js";
import "./directory-runtime-Dv9ukfmB.js";
import "./read-only-account-inspect-gnHd6Zsr.js";
import "./web-media-C4M86664.js";
import "./temp-path-BPp7EkAA.js";
import "./media-understanding-BGz4x5yF.js";
import "./web-media-ChNoP5pI.js";
import "./state-paths-kGWIOBBS.js";
import "./llm-task-CVbsbiq5.js";
import "./pi-model-discovery-BpWH93Dy.js";
import "./exec-inline-eval-BXBKoGM8.js";
import "./target-registry-Che-i6sL.js";
import "./external-content-DD7F5dje.js";
import "./brave-DlJT7GTf.js";
import "./duckduckgo-Bkn2wdAA.js";
import "./exa-DpNIw7fA.js";
import "./security-runtime-kOJWyLps.js";
import "./provider-usage-BWl6XLRe.js";
import "./provider-onboard-DSTfYPeq.js";
import "./perplexity-Diz33fcS.js";
import "./stagger-BkeVxPT8.js";
import "./command-secret-targets-BYcKChcy.js";
import "./delivery-queue-gz62qXSE.js";
import "./channel-summary-wFTF6Nzp.js";
import "./session-system-events-CXg2CZLB.js";
import "./tool-policy-match-BNFjIlBI.js";
import "./runtime-DB0HL3pI.js";
import "./channel-status-BsYjzyR-.js";
import "./discord-core-bSS6QgzK.js";
import "./cli-runtime-hF9d0wpB.js";
import "./config-presence-CovoxGn8.js";
import "./query-expansion-CzcRNQqC.js";
import "./search-manager-PSAQEKfx.js";
import "./acp-runtime-BUMdU4AF.js";
import "./telegram-core-BaYrcVUp.js";
import "./audit-CO95QNfX.js";
import "./gateway-runtime-DhcHrysu.js";
import "./connection-auth-COroqkf4.js";
import "./mcp-config-Dp646rw6.js";
import "./audit-EvVP3T74.js";
import "./thread-bindings.messages-TP9SUuUL.js";
import "./thread-bindings-r8BO6n9H.js";
import "./ui-Ct7fSJa7.js";
import "./src-PBF8suXB.js";
import "./resolve-users-CZ4wEiHp.js";
import "./imessage-CSS2xDUy.js";
import "./process-runtime-8YfSZ5xw.js";
import "./bluebubbles-3Ibx4fJa.js";
import "./setup-browser-DBIhFnBs.js";
import "./setup-runtime-Bg7k8Nqz.js";
import "./setup-tools-DK7F0IQR.js";
import "./setup-group-access-COfUTa3c.js";
import "./bluebubbles-DNSPE918.js";
import "./extension-shared-BLlTRQdf.js";
import "./feishu-Iku76BpW.js";
import "./persistent-dedupe-BCfLRs9-.js";
import "./imessage-core-DoDC70NH.js";
import "./irc-mF4QT_Qw.js";
import "./outbound-media-Ce_HHIge.js";
import "./mattermost-ClneYTJX.js";
import "./nextcloud-talk-nknBqfin.js";
import "./param-readers-BzTNWQBw.js";
import "./slack-core-BhQXgV26.js";
import "./resolve-users-BT7Tg0xJ.js";
import "./action-runtime-CUUtBLgZ.js";
import "./zalo-DMmuqohC.js";
import "./accounts-LRup6u79.js";
import "./webhook-shared-DTdMOor8.js";
import "./runtime-4GHQmi6d.js";
import "./bluebubbles-Bcw7iU-C.js";
import "./setup-surface-CAwunQxH.js";
import "./discord-DSQg50cI.js";
import "./setup-core-CDRAK1RF.js";
import "./shared-DqX1FsVP.js";
import "./setup-entry-7TAovnKZ.js";
import "./reactions-Dnn_Y2e8.js";
import "./feishu-BMX--rGB.js";
import "./runtime-Dsqow0jd.js";
import "./channel-nmbWsbMR.js";
import "./directory.static-CBFI6rcQ.js";
import "./policy-C98bwlHW.js";
import "./send-DkC3ov72.js";
import "./monitor-CPPX9Bc9.js";
import "./runtime-CMhJ7-2d.js";
import "./setup-core-a_FfaAlo.js";
import "./shared--Cht3FtI.js";
import "./imessage-BiYcUR4V.js";
import "./setup-entry-Bqf91aFY.js";
import "./channel-CeWwFj3t.js";
import "./irc-Bk_xwh75.js";
import "./line-BlO1UC-l.js";
import "./channel-shared-Nazli1xM.js";
import "./setup-entry-Dtljol2L.js";
import "./channel-roDNcCcW.js";
import "./mattermost-BeXiD9z8.js";
import "./channel-B5Dgw55f.js";
import "./nextcloud-talk-CtisFbdX.js";
import "./signal-BuamZgtD.js";
import "./setup-core-BoUlSB_z.js";
import "./shared-CpbJMMlV.js";
import "./setup-entry-lqUKgux4.js";
import "./slack-LJkwnb5s.js";
import "./setup-surface-DEymrLZb.js";
import "./setup-entry-DcVCqE1R.js";
import "./channel-BpTUBpgg.js";
import "./synology-chat-CAjNP9sm.js";
import "./telegram-BEVBMePo.js";
import "./shared-BRv5Kn2w.js";
import "./setup-entry-D1Shilkp.js";
import "./runtime-api-DsOv1c75.js";
import "./channel-BNeaciYj.js";
import "./runtime-C-tPhd8L.js";
import "./zalo-Cy3PlcZz.js";
import "./agent-DBwyxk5i.js";
import "./install-source-utils-L4jI7Syn.js";
import "./installs-XLSjUYkq.js";
import "./status-DE_JvlNY.js";
import { n as runCommandWithRuntime } from "./cli-utils-DwDnCzTo.js";
import "./catalog-QCNKy78q.js";
import "./note-DFCwFXVH.js";
import "./clack-prompter-DKnh22on.js";
import "./plugin-install-plan-BSvG_fXq.js";
import "./bundled-sources-nznCMKgA.js";
import "./plugin-install-COFD2k08.js";
import { t as collectOption } from "./helpers-DDHswsdQ.js";
import "./agents.config-D4Vv3Jgq.js";
import "./provider-auth-choice-helpers-BeLNIaNI.js";
import "./provider-auth-guidance-jLdnZwhv.js";
import "./config-validation-c2Hj1g66.js";
import { a as agentsBindCommand, i as agentsAddCommand, n as agentsSetIdentityCommand, o as agentsBindingsCommand, r as agentsDeleteCommand, s as agentsUnbindCommand, t as agentsListCommand } from "./agents-7AOMbU6w.js";
import "./provider-auth-choices-DH43D9k3.js";
import "./provider-wizard-CY3WYX4H.js";
import "./auth-choice-options-CB2V8P60.js";
import "./auth-choice-prompt-CNHrPnt8.js";
import "./provider-auth-choice-1V3Uu8iM.js";
import "./auth-choice.apply-helpers-7ne6PYSl.js";
import "./provider-auth-choice-preference-GTyyR3LQ.js";
import "./auth-choice-Dgd1UnxX.js";
import "./onboard-channels-DY9A3VfC.js";
//#region src/commands/agent-via-gateway.ts
const NO_GATEWAY_TIMEOUT_MS = 2147e6;
function parseTimeoutSeconds(opts) {
	const raw = opts.timeout !== void 0 ? Number.parseInt(String(opts.timeout), 10) : opts.cfg.agents?.defaults?.timeoutSeconds ?? 600;
	if (Number.isNaN(raw) || raw < 0) throw new Error("--timeout must be a non-negative integer (seconds; 0 means no timeout)");
	return raw;
}
function formatPayloadForLog(payload) {
	const parts = resolveSendableOutboundReplyParts({
		text: payload.text,
		mediaUrls: payload.mediaUrls,
		mediaUrl: typeof payload.mediaUrl === "string" ? payload.mediaUrl : void 0
	});
	const lines = [];
	if (parts.text) lines.push(parts.text.trimEnd());
	for (const url of parts.mediaUrls) lines.push(`MEDIA:${url}`);
	return lines.join("\n").trimEnd();
}
async function agentViaGatewayCommand(opts, runtime) {
	const body = (opts.message ?? "").trim();
	if (!body) throw new Error("Message (--message) is required");
	if (!opts.to && !opts.sessionId && !opts.agent) throw new Error("Pass --to <E.164>, --session-id, or --agent to choose a session");
	const cfg = loadConfig();
	const agentIdRaw = opts.agent?.trim();
	const agentId = agentIdRaw ? normalizeAgentId(agentIdRaw) : void 0;
	if (agentId) {
		if (!listAgentIds(cfg).includes(agentId)) throw new Error(`Unknown agent id "${agentIdRaw}". Use "${formatCliCommand("openclaw agents list")}" to see configured agents.`);
	}
	const timeoutSeconds = parseTimeoutSeconds({
		cfg,
		timeout: opts.timeout
	});
	const gatewayTimeoutMs = timeoutSeconds === 0 ? NO_GATEWAY_TIMEOUT_MS : Math.max(1e4, (timeoutSeconds + 30) * 1e3);
	const sessionKey = resolveSessionKeyForRequest({
		cfg,
		agentId,
		to: opts.to,
		sessionId: opts.sessionId
	}).sessionKey;
	const channel = normalizeMessageChannel(opts.channel);
	const idempotencyKey = opts.runId?.trim() || randomIdempotencyKey();
	const response = await withProgress({
		label: "Waiting for agent reply…",
		indeterminate: true,
		enabled: opts.json !== true
	}, async () => await callGateway({
		method: "agent",
		params: {
			message: body,
			agentId,
			to: opts.to,
			replyTo: opts.replyTo,
			sessionId: opts.sessionId,
			sessionKey,
			thinking: opts.thinking,
			deliver: Boolean(opts.deliver),
			channel,
			replyChannel: opts.replyChannel,
			replyAccountId: opts.replyAccount,
			bestEffortDeliver: opts.bestEffortDeliver,
			timeout: timeoutSeconds,
			lane: opts.lane,
			extraSystemPrompt: opts.extraSystemPrompt,
			idempotencyKey
		},
		expectFinal: true,
		timeoutMs: gatewayTimeoutMs,
		clientName: GATEWAY_CLIENT_NAMES.CLI,
		mode: GATEWAY_CLIENT_MODES.CLI
	}));
	if (opts.json) {
		writeRuntimeJson(runtime, response);
		return response;
	}
	const payloads = (response?.result)?.payloads ?? [];
	if (payloads.length === 0) {
		runtime.log(response?.summary ? String(response.summary) : "No reply from agent.");
		return response;
	}
	for (const payload of payloads) {
		const out = formatPayloadForLog(payload);
		if (out) runtime.log(out);
	}
	return response;
}
async function agentCliCommand(opts, runtime, deps) {
	const localOpts = {
		...opts,
		agentId: opts.agent,
		replyAccountId: opts.replyAccount
	};
	if (opts.local === true) return await agentCommand(localOpts, runtime, deps);
	try {
		return await agentViaGatewayCommand(opts, runtime);
	} catch (err) {
		runtime.error?.(`Gateway agent failed; falling back to embedded: ${String(err)}`);
		return await agentCommand(localOpts, runtime, deps);
	}
}
//#endregion
//#region src/cli/program/register.agent.ts
function registerAgentCommands(program, args) {
	program.command("agent").description("Run an agent turn via the Gateway (use --local for embedded)").requiredOption("-m, --message <text>", "Message body for the agent").option("-t, --to <number>", "Recipient number in E.164 used to derive the session key").option("--session-id <id>", "Use an explicit session id").option("--agent <id>", "Agent id (overrides routing bindings)").option("--thinking <level>", "Thinking level: off | minimal | low | medium | high | xhigh").option("--verbose <on|off>", "Persist agent verbose level for the session").option("--channel <channel>", `Delivery channel: ${args.agentChannelOptions} (omit to use the main session channel)`).option("--reply-to <target>", "Delivery target override (separate from session routing)").option("--reply-channel <channel>", "Delivery channel override (separate from routing)").option("--reply-account <id>", "Delivery account id override").option("--local", "Run the embedded agent locally (requires model provider API keys in your shell)", false).option("--deliver", "Send the agent's reply back to the selected channel", false).option("--json", "Output result as JSON", false).option("--timeout <seconds>", "Override agent command timeout (seconds, default 600 or config value)").addHelpText("after", () => `
${theme.heading("Examples:")}
${formatHelpExamples([
		["openclaw agent --to +15555550123 --message \"status update\"", "Start a new session."],
		["openclaw agent --agent ops --message \"Summarize logs\"", "Use a specific agent."],
		["openclaw agent --session-id 1234 --message \"Summarize inbox\" --thinking medium", "Target a session with explicit thinking level."],
		["openclaw agent --to +15555550123 --message \"Trace logs\" --verbose on --json", "Enable verbose logging and JSON output."],
		["openclaw agent --to +15555550123 --message \"Summon reply\" --deliver", "Deliver reply."],
		["openclaw agent --agent ops --message \"Generate report\" --deliver --reply-channel slack --reply-to \"#reports\"", "Send reply to a different channel/target."]
	])}

${theme.muted("Docs:")} ${formatDocsLink("/cli/agent", "docs.openclaw.ai/cli/agent")}`).action(async (opts) => {
		setVerbose((typeof opts.verbose === "string" ? opts.verbose.toLowerCase() : "") === "on");
		const deps = createDefaultDeps();
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentCliCommand(opts, defaultRuntime, deps);
		});
	});
	const agents = program.command("agents").description("Manage isolated agents (workspaces + auth + routing)").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/agents", "docs.openclaw.ai/cli/agents")}\n`);
	agents.command("list").description("List configured agents").option("--json", "Output JSON instead of text", false).option("--bindings", "Include routing bindings", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsListCommand({
				json: Boolean(opts.json),
				bindings: Boolean(opts.bindings)
			}, defaultRuntime);
		});
	});
	agents.command("bindings").description("List routing bindings").option("--agent <id>", "Filter by agent id").option("--json", "Output JSON instead of text", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsBindingsCommand({
				agent: opts.agent,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("bind").description("Add routing bindings for an agent").option("--agent <id>", "Agent id (defaults to current default agent)").option("--bind <channel[:accountId]>", "Binding to add (repeatable). If omitted, accountId is resolved by channel defaults/hooks.", collectOption, []).option("--json", "Output JSON summary", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsBindCommand({
				agent: opts.agent,
				bind: Array.isArray(opts.bind) ? opts.bind : void 0,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("unbind").description("Remove routing bindings for an agent").option("--agent <id>", "Agent id (defaults to current default agent)").option("--bind <channel[:accountId]>", "Binding to remove (repeatable)", collectOption, []).option("--all", "Remove all bindings for this agent", false).option("--json", "Output JSON summary", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsUnbindCommand({
				agent: opts.agent,
				bind: Array.isArray(opts.bind) ? opts.bind : void 0,
				all: Boolean(opts.all),
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("add [name]").description("Add a new isolated agent").option("--workspace <dir>", "Workspace directory for the new agent").option("--model <id>", "Model id for this agent").option("--agent-dir <dir>", "Agent state directory for this agent").option("--bind <channel[:accountId]>", "Route channel binding (repeatable)", collectOption, []).option("--non-interactive", "Disable prompts; requires --workspace", false).option("--json", "Output JSON summary", false).action(async (name, opts, command) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			const hasFlags = hasExplicitOptions(command, [
				"workspace",
				"model",
				"agentDir",
				"bind",
				"nonInteractive"
			]);
			await agentsAddCommand({
				name: typeof name === "string" ? name : void 0,
				workspace: opts.workspace,
				model: opts.model,
				agentDir: opts.agentDir,
				bind: Array.isArray(opts.bind) ? opts.bind : void 0,
				nonInteractive: Boolean(opts.nonInteractive),
				json: Boolean(opts.json)
			}, defaultRuntime, { hasFlags });
		});
	});
	agents.command("set-identity").description("Update an agent identity (name/theme/emoji/avatar)").option("--agent <id>", "Agent id to update").option("--workspace <dir>", "Workspace directory used to locate the agent + IDENTITY.md").option("--identity-file <path>", "Explicit IDENTITY.md path to read").option("--from-identity", "Read values from IDENTITY.md", false).option("--name <name>", "Identity name").option("--theme <theme>", "Identity theme").option("--emoji <emoji>", "Identity emoji").option("--avatar <value>", "Identity avatar (workspace path, http(s) URL, or data URI)").option("--json", "Output JSON summary", false).addHelpText("after", () => `
${theme.heading("Examples:")}
${formatHelpExamples([
		["openclaw agents set-identity --agent main --name \"OpenClaw\" --emoji \"🦞\"", "Set name + emoji."],
		["openclaw agents set-identity --agent main --avatar avatars/openclaw.png", "Set avatar path."],
		["openclaw agents set-identity --workspace ~/.openclaw/workspace --from-identity", "Load from IDENTITY.md."],
		["openclaw agents set-identity --identity-file ~/.openclaw/workspace/IDENTITY.md --agent main", "Use a specific IDENTITY.md."]
	])}
`).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsSetIdentityCommand({
				agent: opts.agent,
				workspace: opts.workspace,
				identityFile: opts.identityFile,
				fromIdentity: Boolean(opts.fromIdentity),
				name: opts.name,
				theme: opts.theme,
				emoji: opts.emoji,
				avatar: opts.avatar,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("delete <id>").description("Delete an agent and prune workspace/state").option("--force", "Skip confirmation", false).option("--json", "Output JSON summary", false).action(async (id, opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsDeleteCommand({
				id: String(id),
				force: Boolean(opts.force),
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.action(async () => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsListCommand({}, defaultRuntime);
		});
	});
}
//#endregion
export { registerAgentCommands };
