// openclaw/dist/shared-DqX1FsVP.js
import { n as DiscordConfigSchema } from "./zod-schema.providers-core-DLPfih2y.js";
import { r as getChatChannelMeta } from "./registry-xyHjVLxh.js";
import { By as inspectDiscordAccount, Gy as resolveDefaultDiscordAccountId, H_ as adaptScopedAccountAccessor, Hy as listDiscordAccountIds, K_ as createScopedChannelConfigAdapter, Ky as resolveDiscordAccount, P_ as createChannelPluginBase, x_ as formatAllowFromLowercase } from "./pi-embedded-CzQCqSlH.js";
import { n as describeAccountSnapshot } from "./account-helpers-BVDd7S3q.js";
import { r as buildChannelConfigSchema } from "./config-schema-Bpwy_blm.js";
import { n as createDiscordSetupWizardProxy } from "./setup-core-CDRAK1RF.js";
//#region extensions/discord/src/shared.ts
const DISCORD_CHANNEL = "discord";
async function loadDiscordChannelRuntime() {
	return await import("./channel.runtime-Cg6o8nxV.js");
}
const discordSetupWizard = createDiscordSetupWizardProxy(async () => (await loadDiscordChannelRuntime()).discordSetupWizard);
const discordConfigAdapter = createScopedChannelConfigAdapter({
	sectionKey: DISCORD_CHANNEL,
	listAccountIds: listDiscordAccountIds,
	resolveAccount: adaptScopedAccountAccessor(resolveDiscordAccount),
	inspectAccount: adaptScopedAccountAccessor(inspectDiscordAccount),
	defaultAccountId: resolveDefaultDiscordAccountId,
	clearBaseFields: ["token", "name"],
	resolveAllowFrom: (account) => account.config.dm?.allowFrom,
	formatAllowFrom: (allowFrom) => formatAllowFromLowercase({ allowFrom }),
	resolveDefaultTo: (account) => account.config.defaultTo
});
function createDiscordPluginBase(params) {
	return createChannelPluginBase({
		id: DISCORD_CHANNEL,
		setupWizard: discordSetupWizard,
		meta: { ...getChatChannelMeta(DISCORD_CHANNEL) },
		capabilities: {
			chatTypes: [
				"direct",
				"channel",
				"thread"
			],
			polls: true,
			reactions: true,
			threads: true,
			media: true,
			nativeCommands: true
		},
		streaming: { blockStreamingCoalesceDefaults: {
			minChars: 1500,
			idleMs: 1e3
		} },
		reload: { configPrefixes: ["channels.discord"] },
		configSchema: buildChannelConfigSchema(DiscordConfigSchema),
		config: {
			...discordConfigAdapter,
			isConfigured: (account) => Boolean(account.token?.trim()),
			describeAccount: (account) => describeAccountSnapshot({
				account,
				configured: Boolean(account.token?.trim()),
				extra: { tokenSource: account.tokenSource }
			})
		},
		setup: params.setup
	});
}
//#endregion
export { discordConfigAdapter as n, createDiscordPluginBase as t };
