// openclaw/dist/plugin-sdk/discord-core.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import { n as DiscordConfigSchema } from "../zod-schema.providers-core-DLPfih2y.js";
import { r as getChatChannelMeta } from "../registry-xyHjVLxh.js";
import "../zod-schema.agent-runtime-C659GPl8.js";
import "../zod-schema.core-yLTNC4-K.js";
import { a as withNormalizedTimestamp } from "../date-time-DCAyaBop.js";
import { c as jsonResult, d as readNumberParam, f as readReactionParams, h as readStringParam, l as parseAvailableTags, p as readStringArrayParam } from "../common-De-BVwIv.js";
import "../image-ops-BLXW6hKm.js";
import "../path-alias-guards-CgP-6a1d.js";
import { t as assertMediaNotDataUrl } from "../sandbox-paths-Bk2ZwMsR.js";
import "../mime-D7Q9o2pi.js";
import { r as buildChannelConfigSchema } from "../config-schema-Bpwy_blm.js";
import "../channel-plugin-common-COvcxNrG.js";
import { r as resolvePollMaxSelections } from "../polls-D9egmDCg.js";
import "../discord-core-bSS6QgzK.js";
export { DiscordConfigSchema, assertMediaNotDataUrl, buildChannelConfigSchema, getChatChannelMeta, jsonResult, parseAvailableTags, readNumberParam, readReactionParams, readStringArrayParam, readStringParam, resolvePollMaxSelections, withNormalizedTimestamp };
