// openclaw/dist/plugin-sdk/media-understanding.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../logger-BmpSCz93.js";
import "../ip-DIgtRRTW.js";
import "../ssrf-rHmQXFZ4.js";
import "../fetch-guard-BH3MjZoA.js";
import { a as describeImagesWithModel, c as postJsonRequest, i as describeImageWithModel, l as postTranscriptionRequest, n as deepgramMediaUnderstandingProvider, o as assertOkOrThrowHttpError, r as transcribeOpenAiCompatibleAudio, s as normalizeBaseUrl, t as groqMediaUnderstandingProvider, u as requireTranscriptionText } from "../media-understanding-BGz4x5yF.js";
export { assertOkOrThrowHttpError, deepgramMediaUnderstandingProvider, describeImageWithModel, describeImagesWithModel, groqMediaUnderstandingProvider, normalizeBaseUrl, postJsonRequest, postTranscriptionRequest, requireTranscriptionText, transcribeOpenAiCompatibleAudio };
