// openclaw/dist/plugin-sdk/provider-auth-api-key.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../utils-CS0Ikux6.js";
import "../paths-BFl2-hCf.js";
import "../auth-profiles-CVFpgeL-.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../agent-scope-DoT9OqaV.js";
import "../model-selection-B6ao45a4.js";
import "../io-cPs4dU7X.js";
import "../host-env-security-DEKL50zA.js";
import "../shell-env-BKuVS72k.js";
import "../safe-text-DIDDfQyI.js";
import "../version-BpHNkJed.js";
import "../env-substitution-B8NFl2Jd.js";
import "../includes-CCK6fRRs.js";
import "../zod-schema.providers-core-DLPfih2y.js";
import "../legacy-web-search-CTvir-Bl.js";
import "../registry-xyHjVLxh.js";
import "../config-state-Dx9-tLPS.js";
import "../min-host-version-xbc6BJ_K.js";
import "../manifest-registry-D5E7Gxgl.js";
import "../runtime-guard-QGt7fm0l.js";
import "../avatar-policy-DVf9B9eu.js";
import "../ip-DIgtRRTW.js";
import "../zod-schema.agent-runtime-C659GPl8.js";
import "../zod-schema.core-yLTNC4-K.js";
import "../config-CB4aYWqd.js";
import "../file-lock-DoKDW8jx.js";
import "../audit-fs-CEN00XrG.js";
import "../resolve-uXRpJb-M.js";
import { n as normalizeSecretInput, t as normalizeOptionalSecretInput } from "../normalize-secret-input-DpCJFpMI.js";
import { a as upsertAuthProfile } from "../profiles-Cyg58FCO.js";
import "../provider-env-vars-Bhzl_gIs.js";
import { a as normalizeSecretInputModeInput, i as normalizeApiKeyInput, n as ensureApiKeyFromOptionEnvOrPrompt, r as formatApiKeyPreview, s as validateApiKeyInput } from "../provider-auth-input-DxjPWsSV.js";
import "../anthropic-vertex-provider-e_sRR8fp.js";
import { t as resolveSecretInputModeForEnvSelection } from "../provider-auth-mode-BBnX4BSi.js";
import { n as promptSecretRefForSetup } from "../provider-auth-ref-D7iRHU8I.js";
import { n as buildApiKeyCredential, t as applyAuthProfileConfig } from "../provider-auth-helpers-BeqTsaJO.js";
import { t as createProviderApiKeyAuthMethod } from "../provider-api-key-auth-C6RcLhw0.js";
import "../provider-auth-api-key-C2k_fzZb.js";
export { applyAuthProfileConfig, buildApiKeyCredential, createProviderApiKeyAuthMethod, ensureApiKeyFromOptionEnvOrPrompt, formatApiKeyPreview, normalizeApiKeyInput, normalizeOptionalSecretInput, normalizeSecretInput, normalizeSecretInputModeInput, promptSecretRefForSetup, resolveSecretInputModeForEnvSelection, upsertAuthProfile, validateApiKeyInput };
