// openclaw/dist/plugin-sdk/sandbox.js
import "../redact-BDinS1q9.js";
import "../errors-BxyFnvP3.js";
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import { n as resolvePreferredOpenClawTmpDir } from "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../utils-CS0Ikux6.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../agent-scope-DoT9OqaV.js";
import "../model-selection-B6ao45a4.js";
import "../io-cPs4dU7X.js";
import "../host-env-security-DEKL50zA.js";
import "../shell-env-BKuVS72k.js";
import "../safe-text-DIDDfQyI.js";
import "../version-BpHNkJed.js";
import "../env-substitution-B8NFl2Jd.js";
import "../includes-CCK6fRRs.js";
import "../zod-schema.providers-core-DLPfih2y.js";
import "../legacy-web-search-CTvir-Bl.js";
import "../registry-xyHjVLxh.js";
import "../config-state-Dx9-tLPS.js";
import "../min-host-version-xbc6BJ_K.js";
import "../manifest-registry-D5E7Gxgl.js";
import "../runtime-guard-QGt7fm0l.js";
import "../avatar-policy-DVf9B9eu.js";
import "../ip-DIgtRRTW.js";
import "../zod-schema.agent-runtime-C659GPl8.js";
import "../zod-schema.core-yLTNC4-K.js";
import "../config-CB4aYWqd.js";
import "../audit-fs-CEN00XrG.js";
import "../resolve-uXRpJb-M.js";
import "../tailscale-CSUwCuE9.js";
import "../tailnet-BVzEE6AW.js";
import "../net-Dk658jWW.js";
import "../auth-C_pTuZtn.js";
import "../credentials-CO55Yx_u.js";
import "../session-write-lock-bE2vQvvO.js";
import "../ports-lsof-C3NOmuKA.js";
import "../ports-CQIuVpXl.js";
import "../tool-catalog-6VPSKPp9.js";
import "../docker-YL21-mX4.js";
import { _ as runSshSandboxCommand, b as createRemoteShellSandboxFsBridge, c as getSandboxBackendManager, d as buildExecRemoteCommand, f as buildRemoteCommand, g as disposeSshSandboxSession, h as createSshSandboxSessionFromSettings, l as registerSandboxBackend, m as createSshSandboxSessionFromConfigText, p as buildSshSandboxArgv, s as getSandboxBackendFactory, u as requireSandboxBackendFactory, v as shellEscape, y as uploadDirectoryToSshTarget } from "../sandbox-C3lR2qEl.js";
import "../image-ops-BLXW6hKm.js";
import "../path-alias-guards-CgP-6a1d.js";
import "../sandbox-paths-Bk2ZwMsR.js";
import "../mime-D7Q9o2pi.js";
import "../ssrf-rHmQXFZ4.js";
import "../fs-safe-Dvzkbqib.js";
import "../frontmatter-DUZeCF_V.js";
import "../env-overrides-DbK0fm-W.js";
import "../skills-CreXENx9.js";
import "../config-BkPtspys.js";
import "../routes-CdaHRCET.js";
import "../ssh-tunnel-CnI7n7iv.js";
import "../server-middleware-DcKqo5DB.js";
import { t as runPluginCommandWithTimeout } from "../run-command-Co6IWubr.js";
import "../sandbox-BKVo5lbD.js";
export { buildExecRemoteCommand, buildRemoteCommand, buildSshSandboxArgv, createRemoteShellSandboxFsBridge, createSshSandboxSessionFromConfigText, createSshSandboxSessionFromSettings, disposeSshSandboxSession, getSandboxBackendFactory, getSandboxBackendManager, registerSandboxBackend, requireSandboxBackendFactory, resolvePreferredOpenClawTmpDir, runPluginCommandWithTimeout, runSshSandboxCommand, shellEscape, uploadDirectoryToSshTarget };
