// openclaw/dist/plugin-sdk/self-hosted-provider-setup.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../utils-CS0Ikux6.js";
import "../paths-BFl2-hCf.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../agent-scope-DoT9OqaV.js";
import "../model-selection-B6ao45a4.js";
import "../file-lock-DoKDW8jx.js";
import "../profiles-Cyg58FCO.js";
import "../provider-catalog-7bUKai5_.js";
import "../provider-env-vars-Bhzl_gIs.js";
import "../provider-model-minimax-XdwKeUOE.js";
import "../provider-models-DfgBoVpf.js";
import "../anthropic-vertex-provider-e_sRR8fp.js";
import { a as SELF_HOSTED_DEFAULT_COST, i as SELF_HOSTED_DEFAULT_CONTEXT_WINDOW, n as buildSglangProvider, o as SELF_HOSTED_DEFAULT_MAX_TOKENS, r as buildVllmProvider } from "../models-config.providers.discovery-Ch23kOpG.js";
import "../provider-auth-helpers-BeqTsaJO.js";
import "../upsert-with-lock-B2QGUxoM.js";
import { a as promptAndConfigureOpenAICompatibleSelfHostedProviderAuth, i as promptAndConfigureOpenAICompatibleSelfHostedProvider, n as configureOpenAICompatibleSelfHostedProviderNonInteractive, r as discoverOpenAICompatibleSelfHostedProvider, t as applyProviderDefaultModel } from "../provider-self-hosted-setup-CdR2owmR.js";
export { SELF_HOSTED_DEFAULT_CONTEXT_WINDOW, SELF_HOSTED_DEFAULT_COST, SELF_HOSTED_DEFAULT_MAX_TOKENS, applyProviderDefaultModel, buildSglangProvider, buildVllmProvider, configureOpenAICompatibleSelfHostedProviderNonInteractive, discoverOpenAICompatibleSelfHostedProvider, promptAndConfigureOpenAICompatibleSelfHostedProvider, promptAndConfigureOpenAICompatibleSelfHostedProviderAuth };
