// openclaw/dist/plugin-sdk/telegram-core.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import { _ as normalizeAccountId } from "../session-key-DAhnzjyr.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import { l as TelegramConfigSchema } from "../zod-schema.providers-core-DLPfih2y.js";
import { r as getChatChannelMeta } from "../registry-xyHjVLxh.js";
import "../zod-schema.agent-runtime-C659GPl8.js";
import "../zod-schema.core-yLTNC4-K.js";
import { c as jsonResult, d as readNumberParam, f as readReactionParams, h as readStringParam, m as readStringOrNumberParam, p as readStringArrayParam } from "../common-De-BVwIv.js";
import "../image-ops-BLXW6hKm.js";
import "../mime-D7Q9o2pi.js";
import { r as buildChannelConfigSchema } from "../config-schema-Bpwy_blm.js";
import "../channel-plugin-common-COvcxNrG.js";
import { r as resolvePollMaxSelections } from "../polls-D9egmDCg.js";
import "../telegram-core-BaYrcVUp.js";
export { TelegramConfigSchema, buildChannelConfigSchema, getChatChannelMeta, jsonResult, normalizeAccountId, readNumberParam, readReactionParams, readStringArrayParam, readStringOrNumberParam, readStringParam, resolvePollMaxSelections };
