// openclaw/dist/extensions/minimax/index.js
import "../../redact-BDinS1q9.js";
import "../../errors-BxyFnvP3.js";
import "../../unhandled-rejections-CDJ8dOVP.js";
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../boolean-C3GkJetE.js";
import "../../env-Dnra1IpT.js";
import "../../warning-filter-CBhOcgHd.js";
import "../../utils-CS0Ikux6.js";
import "../../links-8xRhWBQL.js";
import "../../paths-BFl2-hCf.js";
import "../../auth-profiles-CVFpgeL-.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../agent-scope-DoT9OqaV.js";
import "../../model-selection-B6ao45a4.js";
import "../../io-cPs4dU7X.js";
import "../../host-env-security-DEKL50zA.js";
import "../../shell-env-BKuVS72k.js";
import "../../safe-text-DIDDfQyI.js";
import "../../version-BpHNkJed.js";
import "../../env-substitution-B8NFl2Jd.js";
import "../../includes-CCK6fRRs.js";
import "../../zod-schema.providers-core-DLPfih2y.js";
import "../../legacy-web-search-CTvir-Bl.js";
import "../../registry-xyHjVLxh.js";
import "../../config-state-Dx9-tLPS.js";
import "../../min-host-version-xbc6BJ_K.js";
import "../../manifest-registry-D5E7Gxgl.js";
import "../../runtime-guard-QGt7fm0l.js";
import "../../avatar-policy-DVf9B9eu.js";
import "../../ip-DIgtRRTW.js";
import "../../zod-schema.agent-runtime-C659GPl8.js";
import "../../zod-schema.core-yLTNC4-K.js";
import "../../config-CB4aYWqd.js";
import "../../file-lock-DoKDW8jx.js";
import "../../audit-fs-CEN00XrG.js";
import "../../resolve-uXRpJb-M.js";
import { c as ensureAuthProfileStore, n as listProfilesForProvider } from "../../profiles-Cyg58FCO.js";
import "../../tailscale-CSUwCuE9.js";
import "../../tailnet-BVzEE6AW.js";
import "../../net-Dk658jWW.js";
import "../../auth-C_pTuZtn.js";
import "../../credentials-CO55Yx_u.js";
import "../../message-channel-DUrzQUcI.js";
import "../../store-D_eNuCKK.js";
import "../../runtime-C8dQugND.js";
import "../../plugins-DGdgUNcN.js";
import "../../sessions-C8WesgCl.js";
import "../../paths-BPfguEtH.js";
import "../../session-write-lock-bE2vQvvO.js";
import "../../method-scopes-BWG4Q18M.js";
import "../../call-VZCb020X.js";
import "../../prompt-style-MhFLSlua.js";
import "../../ports-lsof-C3NOmuKA.js";
import "../../restart-stale-pids-CQbYk99d.js";
import "../../ports-CQIuVpXl.js";
import "../../logging-C2L37N2X.js";
import "../../commands-CrPoYX9r.js";
import "../../issue-format-CP-gqjZB.js";
import "../../identity-D8H54Ni5.js";
import "../../heartbeat-FZaB85hC.js";
import { Rh as buildOauthProviderAuthResult } from "../../pi-embedded-CzQCqSlH.js";
import "../../internal-hooks-D4lZfNM5.js";
import "../../multimodal-De3qv_72.js";
import "../../memory-search-DRm4OVlH.js";
import { C as buildMinimaxPortalProvider, w as buildMinimaxProvider } from "../../provider-catalog-7bUKai5_.js";
import "../../secret-input-BNBMqe2K.js";
import "../../bindings-RcwGGOd3.js";
import "../../resolve-route-DbP_XX4E.js";
import "../../routing-B37UwRwq.js";
import "../../identity-file-CCks_qJo.js";
import "../../outbound-runtime-oDjCAP7B.js";
import "../../provider-env-vars-Bhzl_gIs.js";
import "../../provider-auth-input-DxjPWsSV.js";
import { o as isMiniMaxModernModelId, r as MINIMAX_TEXT_MODEL_CATALOG, t as MINIMAX_DEFAULT_MODEL_ID } from "../../provider-model-minimax-XdwKeUOE.js";
import "../../provider-models-DfgBoVpf.js";
import "../../anthropic-vertex-provider-e_sRR8fp.js";
import "../../models-config.providers.discovery-Ch23kOpG.js";
import "../../text-runtime-B9_24WY5.js";
import "../../tool-catalog-6VPSKPp9.js";
import "../../docker-YL21-mX4.js";
import "../../sandbox-C3lR2qEl.js";
import "../../common-De-BVwIv.js";
import "../../image-ops-BLXW6hKm.js";
import "../../thinking-CSfcimRZ.js";
import "../../path-alias-guards-CgP-6a1d.js";
import "../../sandbox-paths-Bk2ZwMsR.js";
import "../../channel-actions-BzT1SAu9.js";
import "../../mime-D7Q9o2pi.js";
import "../../ssrf-rHmQXFZ4.js";
import "../../fetch-guard-BH3MjZoA.js";
import "../../provider-web-search-DYdDI2c9.js";
import "../../manager-CTkU8M9E.js";
import "../../config-schema-Bpwy_blm.js";
import "../../channel-plugin-common-COvcxNrG.js";
import "../../status-helpers-Dd-hr4j4.js";
import "../../conversation-runtime-CqOm9f5F.js";
import "../../runtime-whatsapp-boundary-CWceJs-_.js";
import "../../pairing-store-ClZkWB4v.js";
import "../../json-store-Ju23bywi.js";
import { t as definePluginEntry } from "../../plugin-entry-BNczxv7M.js";
import "../../secret-file-Ctk8Gt6Y.js";
import "../../channel-config-schema-DGuU5OU5.js";
import "../../setup-binary-DXdzcWG_.js";
import "../../archive-Bk8HKpaY.js";
import "../../fs-safe-Dvzkbqib.js";
import "../../signal-cli-install-BJB9N18P.js";
import "../../setup-wizard-proxy-D0W3ulU0.js";
import "../../setup-BYrV6xB7.js";
import "../../dm-policy-shared-DSMJu_ZS.js";
import "../../hook-runtime-DEbp5SYL.js";
import "../../diagnostic-PF4suwLJ.js";
import "../../templating-BueU7auS.js";
import "../../channel-reply-pipeline-C0iCnI2R.js";
import "../../reply-history-BVE7m0XA.js";
import "../../commands-registry.data-Dr5wKe-o.js";
import "../../commands-registry-C-vLUx3f.js";
import "../../frontmatter-DUZeCF_V.js";
import "../../env-overrides-DbK0fm-W.js";
import "../../skills-CreXENx9.js";
import "../../skills-remote-Ctc3qkkj.js";
import "../../workspace-dirs-CKrvld61.js";
import "../../pairing-token-CT1j20jA.js";
import "../../skill-commands-DDonxE3L.js";
import "../../level-overrides-DoL5t4pG.js";
import "../../config-BkPtspys.js";
import "../../routes-CdaHRCET.js";
import "../../ssh-tunnel-CnI7n7iv.js";
import "../../server-middleware-DcKqo5DB.js";
import "../../logging-CSCaAwMS.js";
import "../../config-runtime-ST43sCMq.js";
import "../../exec-approvals-CVzy3Vfp.js";
import "../../webhook-ingress-Chu9Ldhy.js";
import "../../system-events-D9mayVHC.js";
import "../../ssrf-policy-C4n9geuA.js";
import "../../provider-auth-ref-D7iRHU8I.js";
import "../../provider-auth-helpers-BeqTsaJO.js";
import { t as createProviderApiKeyAuthMethod } from "../../provider-api-key-auth-C6RcLhw0.js";
import "../../runtime-env-CjqWVeEh.js";
import "../../pairing-labels-BOG3E444.js";
import "../../directory-runtime-Dv9ukfmB.js";
import "../../read-only-account-inspect-gnHd6Zsr.js";
import "../../web-media-C4M86664.js";
import "../../temp-path-BPp7EkAA.js";
import { a as describeImagesWithModel, i as describeImageWithModel } from "../../media-understanding-BGz4x5yF.js";
import "../../web-media-ChNoP5pI.js";
import "../../state-paths-kGWIOBBS.js";
import "../../llm-task-CVbsbiq5.js";
import "../../pi-model-discovery-BpWH93Dy.js";
import "../../exec-inline-eval-BXBKoGM8.js";
import "../../target-registry-Che-i6sL.js";
import "../../external-content-DD7F5dje.js";
import "../../brave-DlJT7GTf.js";
import "../../duckduckgo-Bkn2wdAA.js";
import "../../exa-DpNIw7fA.js";
import "../../security-runtime-kOJWyLps.js";
import { n as fetchMinimaxUsage } from "../../provider-usage-BWl6XLRe.js";
import { n as applyOnboardAuthAgentModelsAndProviders, t as applyAgentDefaultModelPrimary } from "../../provider-onboarding-config-ChN1HHJG.js";
import "../../provider-onboard-DSTfYPeq.js";
import "../../perplexity-Diz33fcS.js";
import "../../stagger-BkeVxPT8.js";
import "../../command-secret-targets-BYcKChcy.js";
import "../../delivery-queue-gz62qXSE.js";
import "../../channel-summary-wFTF6Nzp.js";
import "../../session-system-events-CXg2CZLB.js";
import "../../tool-policy-match-BNFjIlBI.js";
import "../../runtime-DB0HL3pI.js";
import "../../channel-status-BsYjzyR-.js";
import "../../discord-core-bSS6QgzK.js";
import "../../cli-runtime-hF9d0wpB.js";
import "../../config-presence-CovoxGn8.js";
import "../../query-expansion-CzcRNQqC.js";
import "../../search-manager-PSAQEKfx.js";
import "../../acp-runtime-BUMdU4AF.js";
import "../../telegram-core-BaYrcVUp.js";
import "../../audit-CO95QNfX.js";
import "../../gateway-runtime-DhcHrysu.js";
import "../../connection-auth-COroqkf4.js";
import "../../mcp-config-Dp646rw6.js";
//#region extensions/minimax/media-understanding-provider.ts
const minimaxMediaUnderstandingProvider = {
	id: "minimax",
	capabilities: ["image"],
	describeImage: describeImageWithModel,
	describeImages: describeImagesWithModel
};
const minimaxPortalMediaUnderstandingProvider = {
	id: "minimax-portal",
	capabilities: ["image"],
	describeImage: describeImageWithModel,
	describeImages: describeImagesWithModel
};
//#endregion
//#region extensions/minimax/model-definitions.ts
const MINIMAX_API_BASE_URL = "https://api.minimax.io/anthropic";
const MINIMAX_CN_API_BASE_URL = "https://api.minimaxi.com/anthropic";
`${MINIMAX_DEFAULT_MODEL_ID}`;
const DEFAULT_MINIMAX_CONTEXT_WINDOW = 204800;
const DEFAULT_MINIMAX_MAX_TOKENS = 131072;
const MINIMAX_API_COST = {
	input: .3,
	output: 1.2,
	cacheRead: .06,
	cacheWrite: .375
};
function buildMinimaxModelDefinition(params) {
	const catalog = MINIMAX_TEXT_MODEL_CATALOG[params.id];
	return {
		id: params.id,
		name: params.name ?? catalog?.name ?? `MiniMax ${params.id}`,
		reasoning: params.reasoning ?? catalog?.reasoning ?? false,
		input: ["text"],
		cost: params.cost,
		contextWindow: params.contextWindow,
		maxTokens: params.maxTokens
	};
}
function buildMinimaxApiModelDefinition(modelId) {
	return buildMinimaxModelDefinition({
		id: modelId,
		cost: MINIMAX_API_COST,
		contextWindow: DEFAULT_MINIMAX_CONTEXT_WINDOW,
		maxTokens: DEFAULT_MINIMAX_MAX_TOKENS
	});
}
//#endregion
//#region extensions/minimax/onboard.ts
function applyMinimaxApiProviderConfigWithBaseUrl(cfg, params) {
	const providers = { ...cfg.models?.providers };
	const existingProvider = providers[params.providerId];
	const existingModels = existingProvider?.models ?? [];
	const apiModel = buildMinimaxApiModelDefinition(params.modelId);
	const mergedModels = existingModels.some((model) => model.id === params.modelId) ? existingModels : [...existingModels, apiModel];
	const { apiKey: existingApiKey, ...existingProviderRest } = existingProvider ?? {
		baseUrl: params.baseUrl,
		models: []
	};
	const resolvedApiKey = typeof existingApiKey === "string" ? existingApiKey : void 0;
	const normalizedApiKey = resolvedApiKey?.trim() === "minimax" ? "" : resolvedApiKey;
	providers[params.providerId] = {
		...existingProviderRest,
		baseUrl: params.baseUrl,
		api: "anthropic-messages",
		authHeader: true,
		...normalizedApiKey?.trim() ? { apiKey: normalizedApiKey } : {},
		models: mergedModels.length > 0 ? mergedModels : [apiModel]
	};
	const models = { ...cfg.agents?.defaults?.models };
	const modelRef = `${params.providerId}/${params.modelId}`;
	models[modelRef] = {
		...models[modelRef],
		alias: "Minimax"
	};
	return applyOnboardAuthAgentModelsAndProviders(cfg, {
		agentModels: models,
		providers
	});
}
function applyMinimaxApiConfigWithBaseUrl(cfg, params) {
	return applyAgentDefaultModelPrimary(applyMinimaxApiProviderConfigWithBaseUrl(cfg, params), `${params.providerId}/${params.modelId}`);
}
function applyMinimaxApiConfig(cfg, modelId = "MiniMax-M2.7") {
	return applyMinimaxApiConfigWithBaseUrl(cfg, {
		providerId: "minimax",
		modelId,
		baseUrl: MINIMAX_API_BASE_URL
	});
}
function applyMinimaxApiConfigCn(cfg, modelId = "MiniMax-M2.7") {
	return applyMinimaxApiConfigWithBaseUrl(cfg, {
		providerId: "minimax",
		modelId,
		baseUrl: MINIMAX_CN_API_BASE_URL
	});
}
//#endregion
//#region extensions/minimax/index.ts
const API_PROVIDER_ID = "minimax";
const PORTAL_PROVIDER_ID = "minimax-portal";
const PROVIDER_LABEL = "MiniMax";
const DEFAULT_MODEL = MINIMAX_DEFAULT_MODEL_ID;
const DEFAULT_BASE_URL_CN = "https://api.minimaxi.com/anthropic";
const DEFAULT_BASE_URL_GLOBAL = "https://api.minimax.io/anthropic";
function getDefaultBaseUrl(region) {
	return region === "cn" ? DEFAULT_BASE_URL_CN : DEFAULT_BASE_URL_GLOBAL;
}
function apiModelRef(modelId) {
	return `${API_PROVIDER_ID}/${modelId}`;
}
function portalModelRef(modelId) {
	return `${PORTAL_PROVIDER_ID}/${modelId}`;
}
function buildPortalProviderCatalog(params) {
	return {
		...buildMinimaxPortalProvider(),
		baseUrl: params.baseUrl,
		apiKey: params.apiKey
	};
}
function resolveApiCatalog(ctx) {
	const apiKey = ctx.resolveProviderApiKey(API_PROVIDER_ID).apiKey;
	if (!apiKey) return null;
	return { provider: {
		...buildMinimaxProvider(),
		apiKey
	} };
}
function resolvePortalCatalog(ctx) {
	const explicitProvider = ctx.config.models?.providers?.[PORTAL_PROVIDER_ID];
	const envApiKey = ctx.resolveProviderApiKey(PORTAL_PROVIDER_ID).apiKey;
	const hasProfiles = listProfilesForProvider(ensureAuthProfileStore(ctx.agentDir, { allowKeychainPrompt: false }), PORTAL_PROVIDER_ID).length > 0;
	const explicitApiKey = typeof explicitProvider?.apiKey === "string" ? explicitProvider.apiKey.trim() : void 0;
	const apiKey = envApiKey ?? explicitApiKey ?? (hasProfiles ? "minimax-oauth" : void 0);
	if (!apiKey) return null;
	return { provider: buildPortalProviderCatalog({
		baseUrl: (typeof explicitProvider?.baseUrl === "string" ? explicitProvider.baseUrl.trim() : void 0) || DEFAULT_BASE_URL_GLOBAL,
		apiKey
	}) };
}
function createOAuthHandler(region) {
	const defaultBaseUrl = getDefaultBaseUrl(region);
	const regionLabel = region === "cn" ? "CN" : "Global";
	return async (ctx) => {
		const progress = ctx.prompter.progress(`Starting MiniMax OAuth (${regionLabel})…`);
		try {
			const { loginMiniMaxPortalOAuth } = await import("../../oauth.runtime-B3RcTXaK.js");
			const result = await loginMiniMaxPortalOAuth({
				openUrl: ctx.openUrl,
				note: ctx.prompter.note,
				progress,
				region
			});
			progress.stop("MiniMax OAuth complete");
			if (result.notification_message) await ctx.prompter.note(result.notification_message, "MiniMax OAuth");
			const baseUrl = result.resourceUrl || defaultBaseUrl;
			return buildOauthProviderAuthResult({
				providerId: PORTAL_PROVIDER_ID,
				defaultModel: portalModelRef(DEFAULT_MODEL),
				access: result.access,
				refresh: result.refresh,
				expires: result.expires,
				configPatch: {
					models: { providers: { [PORTAL_PROVIDER_ID]: {
						baseUrl,
						models: []
					} } },
					agents: { defaults: { models: {
						[portalModelRef("MiniMax-M2")]: { alias: "minimax-m2" },
						[portalModelRef("MiniMax-M2.1")]: { alias: "minimax-m2.1" },
						[portalModelRef("MiniMax-M2.1-highspeed")]: { alias: "minimax-m2.1-highspeed" },
						[portalModelRef("MiniMax-M2.7")]: { alias: "minimax-m2.7" },
						[portalModelRef("MiniMax-M2.7-highspeed")]: { alias: "minimax-m2.7-highspeed" },
						[portalModelRef("MiniMax-M2.5")]: { alias: "minimax-m2.5" },
						[portalModelRef("MiniMax-M2.5-highspeed")]: { alias: "minimax-m2.5-highspeed" },
						[portalModelRef("MiniMax-M2.5-Lightning")]: { alias: "minimax-m2.5-lightning" }
					} } }
				},
				notes: [
					"MiniMax OAuth tokens auto-refresh. Re-run login if refresh fails or access is revoked.",
					`Base URL defaults to ${defaultBaseUrl}. Override models.providers.${PORTAL_PROVIDER_ID}.baseUrl if needed.`,
					...result.notification_message ? [result.notification_message] : []
				]
			});
		} catch (err) {
			const errorMsg = err instanceof Error ? err.message : String(err);
			progress.stop(`MiniMax OAuth failed: ${errorMsg}`);
			await ctx.prompter.note("If OAuth fails, verify your MiniMax account has portal access and try again.", "MiniMax OAuth");
			throw err;
		}
	};
}
var minimax_default = definePluginEntry({
	id: API_PROVIDER_ID,
	name: "MiniMax",
	description: "Bundled MiniMax API-key and OAuth provider plugin",
	register(api) {
		api.registerProvider({
			id: API_PROVIDER_ID,
			label: PROVIDER_LABEL,
			docsPath: "/providers/minimax",
			envVars: ["MINIMAX_API_KEY"],
			auth: [createProviderApiKeyAuthMethod({
				providerId: API_PROVIDER_ID,
				methodId: "api-global",
				label: "MiniMax API key (Global)",
				hint: "Global endpoint - api.minimax.io",
				optionKey: "minimaxApiKey",
				flagName: "--minimax-api-key",
				envVar: "MINIMAX_API_KEY",
				promptMessage: "Enter MiniMax API key (sk-api- or sk-cp-)\nhttps://platform.minimax.io/user-center/basic-information/interface-key",
				profileId: "minimax:global",
				allowProfile: false,
				defaultModel: apiModelRef(DEFAULT_MODEL),
				expectedProviders: ["minimax"],
				applyConfig: (cfg) => applyMinimaxApiConfig(cfg),
				wizard: {
					choiceId: "minimax-global-api",
					choiceLabel: "MiniMax API key (Global)",
					choiceHint: "Global endpoint - api.minimax.io",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				}
			}), createProviderApiKeyAuthMethod({
				providerId: API_PROVIDER_ID,
				methodId: "api-cn",
				label: "MiniMax API key (CN)",
				hint: "CN endpoint - api.minimaxi.com",
				optionKey: "minimaxApiKey",
				flagName: "--minimax-api-key",
				envVar: "MINIMAX_API_KEY",
				promptMessage: "Enter MiniMax CN API key (sk-api- or sk-cp-)\nhttps://platform.minimaxi.com/user-center/basic-information/interface-key",
				profileId: "minimax:cn",
				allowProfile: false,
				defaultModel: apiModelRef(DEFAULT_MODEL),
				expectedProviders: ["minimax", "minimax-cn"],
				applyConfig: (cfg) => applyMinimaxApiConfigCn(cfg),
				wizard: {
					choiceId: "minimax-cn-api",
					choiceLabel: "MiniMax API key (CN)",
					choiceHint: "CN endpoint - api.minimaxi.com",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				}
			})],
			catalog: {
				order: "simple",
				run: async (ctx) => resolveApiCatalog(ctx)
			},
			resolveUsageAuth: async (ctx) => {
				const apiKey = ctx.resolveApiKeyFromConfigAndStore({ envDirect: [ctx.env.MINIMAX_CODE_PLAN_KEY, ctx.env.MINIMAX_API_KEY] });
				return apiKey ? { token: apiKey } : null;
			},
			isModernModelRef: ({ modelId }) => isMiniMaxModernModelId(modelId),
			fetchUsageSnapshot: async (ctx) => await fetchMinimaxUsage(ctx.token, ctx.timeoutMs, ctx.fetchFn)
		});
		api.registerProvider({
			id: PORTAL_PROVIDER_ID,
			label: PROVIDER_LABEL,
			docsPath: "/providers/minimax",
			envVars: ["MINIMAX_OAUTH_TOKEN", "MINIMAX_API_KEY"],
			catalog: { run: async (ctx) => resolvePortalCatalog(ctx) },
			auth: [{
				id: "oauth",
				label: "MiniMax OAuth (Global)",
				hint: "Global endpoint - api.minimax.io",
				kind: "device_code",
				wizard: {
					choiceId: "minimax-global-oauth",
					choiceLabel: "MiniMax OAuth (Global)",
					choiceHint: "Global endpoint - api.minimax.io",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				},
				run: createOAuthHandler("global")
			}, {
				id: "oauth-cn",
				label: "MiniMax OAuth (CN)",
				hint: "CN endpoint - api.minimaxi.com",
				kind: "device_code",
				wizard: {
					choiceId: "minimax-cn-oauth",
					choiceLabel: "MiniMax OAuth (CN)",
					choiceHint: "CN endpoint - api.minimaxi.com",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				},
				run: createOAuthHandler("cn")
			}],
			isModernModelRef: ({ modelId }) => isMiniMaxModernModelId(modelId)
		});
		api.registerMediaUnderstandingProvider(minimaxMediaUnderstandingProvider);
		api.registerMediaUnderstandingProvider(minimaxPortalMediaUnderstandingProvider);
	}
});
//#endregion
export { minimax_default as default };
